<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.335">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Trevor Harrington">

<title>Replication_Trevor_Harrington</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="Replication_Trevor_Harrington_files/libs/clipboard/clipboard.min.js"></script>
<script src="Replication_Trevor_Harrington_files/libs/quarto-html/quarto.js"></script>
<script src="Replication_Trevor_Harrington_files/libs/quarto-html/popper.min.js"></script>
<script src="Replication_Trevor_Harrington_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Replication_Trevor_Harrington_files/libs/quarto-html/anchor.min.js"></script>
<link href="Replication_Trevor_Harrington_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Replication_Trevor_Harrington_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Replication_Trevor_Harrington_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Replication_Trevor_Harrington_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Replication_Trevor_Harrington_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Replication_Trevor_Harrington</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Trevor Harrington </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="chapter-1-from-statistical-analysis-to-bioinformatics" class="level1">
<h1>Chapter 1 – From Statistical Analysis to Bioinformatics</h1>
<section id="introduction-sampling-processing-and-analyzing-genomic-data" class="level2">
<h2 class="anchored" data-anchor-id="introduction-sampling-processing-and-analyzing-genomic-data">Introduction: Sampling, Processing, and Analyzing Genomic data</h2>
<p>The goal of this notebook is to complete a series of challenges that will lay the foundation for using RStudio to analyze genomic data. By learning how to effectively store objects in variables, build functions to analyze genetic data, and implement “for” loops when necessary, we aim to enhance our ability to analyze and interpret genomic data more efficiently. Through analyzing genomic data with the tools and concepts covered in the course, we will gain a deeper understanding of genetic information and its crucial role in biological processes. Ultimately, our aim is to leave this course with a well-documented online environment that showcases our accomplishments in analyzing genomic data</p>
</section>
<section id="objectives" class="level2">
<h2 class="anchored" data-anchor-id="objectives">Objectives:</h2>
<ul>
<li><p>Use the arrow (&lt;-) operator to store objects in variables.</p></li>
<li><p>Build functions to process and analyze genetic data.</p></li>
<li><p>Create a list of nucleotides</p></li>
<li><p>Use set.seed() to set a seed for a random number generator to make your results reproducible over time and across different machines (and different researchers too).</p></li>
<li><p>Use R’s sample() function to create a random “genome” for the purpose of testing functions.</p></li>
<li><p>Use paste() with the parameter setting collapse = “” to collapse your random genome into a single long string of nucleotides.</p></li>
<li><p>Learn about <code>for loops</code> and implement them when necessary.</p></li>
<li><p>Read in a real bacterial genome from a text file.</p></li>
<li><p>Use programmatic flow control in the form of if/else if/else statements to run code only when a particular condition is satisfied.</p></li>
<li><p>Solve a bioinformatics problem on the Rosalind platform.</p></li>
</ul>
<hr>
<section id="generating-a-new-variable-containing-each-nucleotide" class="level3">
<h3 class="anchored" data-anchor-id="generating-a-new-variable-containing-each-nucleotide">1.1 Generating a New Variable Containing each Nucleotide</h3>
<ul>
<li><p><strong>Vectors</strong> are a one-dimensional array of values <u>of the same data type</u>.</p>
<ul>
<li><p><code>vector &lt;- c(1, 2, 3, 4)</code> , <code>vector &lt;- 1:4</code> are methods of generating a vector array.</p></li>
<li><p>This code can be accessed by positional metrics, where position is a numerical value. For example, <code>vector[3]</code> will return the value 3 (considering the examples above), or whatever value is located in the third position.</p></li>
</ul></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a new variable containing a vector of four characters for the four nucleotide abbreviations (Adenine, Cytosine, Guanine, Thymine) </span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Vector is  a one-dimensional array-like object that can contain values of the same basic type of data (numerical,character,logical)</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>nt_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>nt_names <span class="co"># simply displays the vectors without organizing</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "T" "G" "C"</code></pre>
</div>
</div>
<hr>
</section>
<section id="generating-a-randomized-string-of-15-nucleotides-variables" class="level3">
<h3 class="anchored" data-anchor-id="generating-a-randomized-string-of-15-nucleotides-variables">1.2 Generating a Randomized String of 15 Nucleotides variables</h3>
<p>Using the vector array created in the previous</p>
<ul>
<li><code>set.seed</code> is used to generate a reproducible algorithm for repeating the</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># `sample` function used to generate a random sample of nucleotides from the nucleotides (nt_name) vector. The `size` argument specifies the number of samples to generate (15)  </span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="do">##`replace = TRUE` indicates sample should be replaced in the pool after being selected (same nucleotide can appear multiple times in the sample)</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="do">## nt_sample calls on vector generated in challenge 1 and stored in the environment</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>rnd_genome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nt_names, <span class="at">size =</span> <span class="dv">15</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(rnd_genome, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CGGAACTCCCAACGC"</code></pre>
</div>
</div>
</section>
<section id="expanding-randomized-string-to-1500-nucleotides" class="level3">
<h3 class="anchored" data-anchor-id="expanding-randomized-string-to-1500-nucleotides">1.3 Expanding Randomized String to 1500 Nucleotides</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a new variable containing a vector of four characters for the four nucleotide abbreviations (Adenine, Cytosine, Guanine, Thymine) </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="do">## Vector is  a one-dimensional array-like object that can contain values of the same basic type of data (numerical,character,logical)</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>nt_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># generate a new value representing the number of nucleotides that will be randomly generated for this dataset</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="do">##This step is not required (numerical value can be used in 'sample') but it is better to call on a value then hard code one into a function. More flexible when df is larger</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>nt_num <span class="ot">&lt;-</span> <span class="dv">1500</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># for reproducibility, using the same seed each time means the "randomness" can be recreated.</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co">#'sample' function randomly picks variables from nt_names vector to generate a new vector n length, in this case n = nt_num (1500)</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>nt_sample <span class="ot">&lt;-</span> <span class="fu">sample</span>(nt_names, <span class="at">size =</span> nt_num, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co"># 'paste' combines the individual nucleotides into a single string rather then a list of individual letters</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="co"># 'collapse' removes the seperator generated between each nucleotide as a result of the 'sample' function interavting with the nucleotides.</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(nt_sample, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CGGAACTCCCAACGCCTTGATTCCCGAGTTCTAAGCCGGATCATTGTGGTTTTTGATTGAGAGTCAATCTCAAACGACGTAAGTAGTGTGCGTTGAGCTCTCGCGGATAGGACTATACCGGACGCGAGTTAAGACTCTGAGACGAAAAATAAGCAGGCCTCTCACTGTCGGTCTTAACTACCCCCACTTCCCGTCGTACATCCACGGTTTCTTAATTCCGTGAACCGTGGTACCATGCCTCACGTATGAAAGAGGTATGAGACGCACACTATCTCCTAGTCACTCGATATAGGCAGGTACCGGACGCTAGTGCATGATTGGGGCAGGTAATTTTCCTCGCCGTTTTGTCGGTTGGACGTTAAGGCGCCCCATACTGGCACCCCAAAGACGCGAATCCAAGCAATTCATCCAGTCGATGGGAAGGCGTATACGCACGCGCGATCTCATATTAAGACGTATCCAGGTACTAACAAAAGCCGTTGGCCCGCAACTTCCGATAGCAACTGAGACAACCATCCGTGGGTATACAATTCATCTGGCCTGCTTTTTCCAGTGCATGGGGGGGCGGGGAATTACTAGTGCCTGCACGCCAGCTTCTTGGTACCCCCGGCTATGGTTTCTTACGAAACATACATCAGTGCGTTACCTTGGCTAAACGGTTTCGTGAAGGACGTGGCGTGGCAAGTGCGCGGGTTACATCCGGGTAGACCGAGCCACAAGAAATAGGTAAACCGGGAATCAGCGTGGTATACGCATAGTCTGTATCCTTCGGGGGGTGATCATTGAACCTTCAGCACGCTCGAGCAGTGCATTGGGTTAGTCCCGGGCTTCTCCATTCCGCAACGGGACTGGTTCCACGGGACGTTACGATGATATCGTGGGAGGCCAACAAGCGCTTGTGACAAAGTTCTGCGGCTGAAACTCGCCATGCGTCTCCTCGCTACCCGTTACACCGGCAAAGCCTAAGAGTTATTTCACTCACCGATCTTCAGACTCTTTAAACGCACTTCTTAATAGCATCCTATTCGGACGGCAATTTTATTCTAACTATTATCCAGGCTCTTAGCATCCCCGAGTTGTTCATGTGTCTCTGAAACAAGTCAAGTCAGTCATGGCTTGCGCCAGGTGAAAAACGAACTTTCCCTTAGTGTCATTAAGCGCAGCTGGACGTGGCCGATCCACGTTCGTTTGGCGAGTAGAACCCAAACTGTCGCATTAACTAGTATCATTATAAGGTGAATGGGGTAGTTATGATCGGTGTTATAAGTACGGAAGGAGCGCTGGGTGAAACTAGGGTTGTACCAAGTTGCCCCACTATGGGCAAAGAGGCTATACGCGTATAGAGTTAGACAGCTGGTAGCGAAGGGTATAATGTCCCCCTGGTAGGAAGATTTCCGTATTCCCACTATATTGTGGGCTACTATTGGTTAAAGTAGCTGGTGGCTAATGCCAAAACCACCGTACGCTCAGGCAAAGTAGGATTAGAGAGATGCTGAAACT"</code></pre>
</div>
</div>
<hr>
</section>
<section id="counting-each-dna-nucleotide-in-a-randomized-dataset-and-using-loop" class="level3">
<h3 class="anchored" data-anchor-id="counting-each-dna-nucleotide-in-a-randomized-dataset-and-using-loop">1.4 Counting each DNA Nucleotide in a Randomized Dataset and using Loop()</h3>
<section id="generating-a-100-nucleotide-strand-and-counting-each-frequency" class="level4">
<h4 class="anchored" data-anchor-id="generating-a-100-nucleotide-strand-and-counting-each-frequency">Generating a 100-nucleotide strand and counting each frequency</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a seed value to ensure that the random numbers generated are reproducible</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a random genome sequence of length 100 by sampling (with replacement) from the nucleotide abbreviation vector still in the environment</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>rnd_genome <span class="ot">&lt;-</span> <span class="fu">sample</span>(nt_names, <span class="at">size =</span> <span class="dv">100</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># 'table' function counts the frequency of each vector (nucleotide) in the random genome sequence and generates a table of the results </span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(rnd_genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>rnd_genome
 A  C  G  T 
23 23 25 29 </code></pre>
</div>
</div>
</section>
<section id="using-loop-function-to-repeat-an-operation" class="level4">
<h4 class="anchored" data-anchor-id="using-loop-function-to-repeat-an-operation">Using Loop() function to repeat an operation</h4>
<ul>
<li><code>for loop</code> functions are widely used and incredibly versatile for repeating an operation with minimal inputs.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create 'loop_result' variable used to store the values being added in each iteration of the loop.</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="do">## loop is initialized to 0 (or whatever value entered for 'loop_result') and then incremented by the current value of i (and the function applied to the loop) in each iteration of the loop</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>loop_result <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># for loop is a control flow statement in programming that allows you to execute a block of code repeatedly.</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="do">## The loop consists of an initialization statement, a test condition, an update statement, and a loop body. In this code</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Here, a loop is used to iterate over a sequence of values from 1 to 10, adding the value 'i' for each iteration.</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>){</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  loop_result <span class="ot">&lt;-</span> loop_result <span class="sc">+</span> i</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(loop_result)</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1
[1] 3
[1] 6
[1] 10
[1] 15
[1] 21
[1] 28
[1] 36
[1] 45
[1] 55</code></pre>
</div>
</div>
<hr>
</section>
</section>
<section id="using-loop-functions" class="level3">
<h3 class="anchored" data-anchor-id="using-loop-functions">1.5 Using loop() functions</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co">#initialize a new dataframe with a value of 1</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>myProduct <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Use "for loop" to iterate (multiply) myProduct by values of j(1-15)</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="do">##kind of confusing, but the iteration variable "j" </span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">15</span>) {</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  myProduct <span class="ot">&lt;-</span> myProduct <span class="sc">*</span> j   <span class="co"># Update myProduct by multiplying with j</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(myProduct)            <span class="co"># Print the updated value of myProduct</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1
[1] 2
[1] 6
[1] 24
[1] 120
[1] 720
[1] 5040
[1] 40320
[1] 362880
[1] 3628800
[1] 39916800
[1] 479001600
[1] 6227020800
[1] 87178291200
[1] 1.307674e+12</code></pre>
</div>
</div>
<hr>
</section>
<section id="combining-sample-paste-and-loop-functions-to-produce-a-set-of-individual-nucleotides" class="level3">
<h3 class="anchored" data-anchor-id="combining-sample-paste-and-loop-functions-to-produce-a-set-of-individual-nucleotides">1.6: Combining sample(), paste() and loop() functions to produce a set of individual nucleotides</h3>
<section id="generating-random-genomic-data-in-a-collapsed-string" class="level4">
<h4 class="anchored" data-anchor-id="generating-random-genomic-data-in-a-collapsed-string">Generating Random Genomic Data in a collapsed string</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>nt <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>genomeLength <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>rnd_genome <span class="ot">&lt;-</span> <span class="fu">paste</span>(</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample</span>(nt, <span class="at">size =</span> genomeLength, <span class="at">replace =</span> <span class="cn">TRUE</span>),</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(rnd_genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TGGAATCTTT"</code></pre>
</div>
</div>
</section>
<section id="counting-adenine-using-loop-function-with-ifelse-statements" class="level4">
<h4 class="anchored" data-anchor-id="counting-adenine-using-loop-function-with-ifelse-statements">Counting Adenine using loop() function with If/else statements</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># "For loop" is used to perform iterations 1:n where n (nchar) refers to the number of characters (variables, nucleotides) in the rnd_genome dataset</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rnd_genome))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  {</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i))</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  }</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "T"
[1] "G"
[1] "G"
[1] "A"
[1] "A"
[1] "T"
[1] "C"
[1] "T"
[1] "T"
[1] "T"</code></pre>
</div>
</div>
</section>
</section>
<section id="modifying-output-to-sum-the-for-loop-results-output" class="level3">
<h3 class="anchored" data-anchor-id="modifying-output-to-sum-the-for-loop-results-output">1.7: Modifying Output to sum the For Loop() Results Output</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set a variable called `sum_A` equal to 0</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>sum_A <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Use a for loop to iterate over each character in the `rnd_genome` variable</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rnd_genome)) {</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># # String subset is checked for each iteration, start = i and end = i means this substring will only extract a single character, i, for each iteration</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="do">## == "A" checks each substring, and only extracts A for each nucleotide in the datafrom</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>) {</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If it is, add 1 to the `sum_A` variable</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    sum_A <span class="ot">&lt;-</span> sum_A <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a named vector called `result` with one element, the number of adenine bases found</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Adenine:"</span> <span class="ot">=</span> sum_A)</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Use the `print` function to output the `result` vector to the console</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(result)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Adenine: 
       2 </code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>sum_A <span class="ot">&lt;-</span> <span class="dv">0</span>  <span class="co"># Create a new variable to store sum of "A" characters extracted from the loop</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rnd_genome)) {</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>) {</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>    sum_A <span class="ot">&lt;-</span> sum_A <span class="sc">+</span> <span class="dv">1</span>   <span class="co"># Adds 1 to sum_A variable when a iterations passes the if "A" statement</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Total number of Adenisine Nucleotides: "</span>, sum_A)  <span class="co"># Print the final value of sum_A using cat() "concatenate and print" function to merge the output of sum_A</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Total number of Adenisine Nucleotides:  2</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Without cat(), print() produces two lines</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="adding-the-remaining-nucleotides-to-the-count" class="level3">
<h3 class="anchored" data-anchor-id="adding-the-remaining-nucleotides-to-the-count">1.8: Adding the remaining nucleotides to the count</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize count variables for each nucleotide to 0</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>sum_A <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>sum_C <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>sum_G <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>sum_T <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Loop through each character in the random genome sequence</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rnd_genome)) {</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Check if the character is "A", "C", "G", or "T", and increment the corresponding count variable if it is</span></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>  <span class="do">##'str_sub' substring function allows each vector in the variable to be checked and summed independent of the other vectors in the string</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>  <span class="do">##'start = i' and 'end = i' is starting at position i and ending at position i for each iteration extracting the "i-th" character of the string, which is then compared to the nucleotide letters A, C, G, and T and added to the count.</span></span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>) {</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>    sum_A <span class="ot">&lt;-</span> sum_A <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>) {</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>    sum_C <span class="ot">&lt;-</span> sum_C <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>) {</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>    sum_G <span class="ot">&lt;-</span> sum_G <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rnd_genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>) {</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>    sum_T <span class="ot">&lt;-</span> sum_T <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a named vector of the counts for each nucleotide</span></span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Adenine:"</span> <span class="ot">=</span> sum_A, <span class="st">"Cytosine:"</span> <span class="ot">=</span> sum_C, <span class="st">"Guanine:"</span> <span class="ot">=</span> sum_G, <span class="st">"Thymine:"</span> <span class="ot">=</span> sum_T)</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the results</span></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(result)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> Adenine: Cytosine:  Guanine:  Thymine: 
        2         1         2         5 </code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="understanding-the-biology-dna-replication-lecture" class="level1">
<h1>Understanding the biology – DNA Replication Lecture</h1>
<section id="useful-terminology" class="level2">
<h2 class="anchored" data-anchor-id="useful-terminology">Useful Terminology</h2>
<ul>
<li><p>Scales of genetic data; genome -&gt; Chromosome -&gt; Gene -&gt; Nucleotide</p>
<ul>
<li><p>Humans have 23 pairs of chromosomes</p></li>
<li><p>Bacteria have a single looping strand containing all of their genetic data</p>
<ul>
<li>Bacteria also contain small, circular DNA molecules called <strong>Plasmids</strong>, used for storing additional data and conjugation/cloning</li>
</ul></li>
</ul></li>
<li><p><strong>Complementary base pairing:</strong> A &amp; T + C &amp; G</p>
<ul>
<li><p>antiparellel strands: DNA always read from <strong>5’ -&gt; 3’</strong></p></li>
<li><p>DNA strand = phosphate group + hydroxyl (sugar) group + base group (A,C,G,T)</p>
<ul>
<li><p>Sugar-Phosphate backbone: creates the 5’ -&gt; 3’ rule (placement of carbon in sugar)</p></li>
<li><p>base pairs held together by matching <strong>Hydrogen bonds.</strong></p></li>
</ul></li>
</ul></li>
<li><p><strong>DNA Polymerase:</strong> Along with a number of other enzymes used to read and create a RNA copy of the DNA.</p>
<h2 id="origins-of-replication" class="anchored"><strong>Origins of Replication</strong></h2>
<p>shorts sequences of genes that can be read to initiate the reading of a gene.</p></li>
<li><p>Many origins of replication in the human genome, where genes are constantly being replicated</p></li>
<li><p>Bacteria (tend to) only have a single origin of replication that travels in both directions</p>
<ul>
<li><p><strong>DnaA Boxes:</strong> specific bacterial marker that identifies the origin for enzymes to begin replication</p>
<ul>
<li>specific sequence used to initiate a DnaA box</li>
</ul></li>
</ul></li>
</ul>
<hr>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>vib_c <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"C:/School/23SPDAY/Biostatistics/Group_Project/Genomes/VibrioCholerae.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>sum_A <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>sum_C <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>sum_G <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>sum_T <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(vib_c)) {</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>) {sum_A <span class="ot">&lt;-</span> sum_A <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>) {sum_C <span class="ot">&lt;-</span> sum_C <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>) {sum_G <span class="ot">&lt;-</span> sum_G <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(vib_c, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>) {sum_T <span class="ot">&lt;-</span> sum_T <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Adenine:"</span> <span class="ot">=</span> sum_A, <span class="st">"Cytosine:"</span> <span class="ot">=</span> sum_C, <span class="st">"Guanine:"</span> <span class="ot">=</span> sum_G, <span class="st">"Thymine:"</span> <span class="ot">=</span> sum_T)</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(result)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> Adenine: Cytosine:  Guanine:  Thymine: 
   293942    263573    256024    294711 </code></pre>
</div>
</div>
<hr>
<section id="rosalind-testing-data" class="level3">
<h3 class="anchored" data-anchor-id="rosalind-testing-data">Rosalind Testing Data</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>rosalind_string <span class="ot">&lt;-</span> <span class="st">"AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGC"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sum_A <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>sum_C <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>sum_G <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>sum_T <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(rosalind_string)) {</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_string, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"A"</span>) {sum_A <span class="ot">&lt;-</span> sum_A <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_string, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>) {sum_C <span class="ot">&lt;-</span> sum_C <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_string, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>) {sum_G <span class="ot">&lt;-</span> sum_G <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">else</span> <span class="cf">if</span>(<span class="fu">str_sub</span>(rosalind_string, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"T"</span>) {sum_T <span class="ot">&lt;-</span> sum_T <span class="sc">+</span> <span class="dv">1</span>}</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a><span class="fu">paste</span>(sum_A,sum_C,sum_G,sum_T)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "20 12 17 21"</code></pre>
</div>
</div>
<ul>
<li>This result matches the key provided on the website, so I assume that this code is effective!</li>
</ul>
<hr>
</section>
</section>
</section>
<section id="chapter-2-bioinformatic-tools-pt.-2" class="level1">
<h1>Chapter 2: Bioinformatic Tools pt.&nbsp;2</h1>
<section id="introduction-improving-tools-and-efficiency" class="level2">
<h2 class="anchored" data-anchor-id="introduction-improving-tools-and-efficiency">Introduction: Improving Tools and Efficiency</h2>
<ul>
<li>Genomes are large, and it is important that the tools for analyzing them are used properly to both validate that the results are accurate, and that as few unnecessary steps as possible are made since they will be time consuming when working with a large string of nucleotides.</li>
</ul>
</section>
<section id="objectives-1" class="level2">
<h2 class="anchored" data-anchor-id="objectives-1">Objectives:</h2>
<ul>
<li><p>Learn to eliminate “single-use code” by encapsulating code in functions which can be easily applied to different genomes.</p></li>
<li><p>Expand your ability to count the frequency of each individual nucleotide within a genome to an ability to count the appearance of patterns in the genome.</p></li>
<li><p>Build and execute functions to find “hidden messages” within the genome by identifying patterns appearing much more often than they would be expected to if a genome included nucleotides at random.</p></li>
<li><p>Exploit transcription errors to narrow a search region for the replication origin.</p></li>
</ul>
<hr>
<section id="generating-functions-for-functions" class="level3">
<h3 class="anchored" data-anchor-id="generating-functions-for-functions">2.0: Generating Functions for Functions</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Functions allow code to be ran without having to enter or copy and paste the code into a new block</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="do">## within function(), the variables that can be altered are added (and pop up when entering in the function)</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="do">## nucleotide = "A" means IF no nucleotide is supplied when using the function, it defaults to "A"</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>nucleotide_frequency <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, <span class="at">nucleotide =</span> <span class="st">"A"</span>){</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="dv">0</span> <span class="co"># generate a new variable for the function to write to</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>  <span class="do">## for each iteration of i from 1 to the end of 'genomestring', if character in the string is the specificied nucleotide, count is updated by +1</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="do">## {} are layers to the formula, which are only run in each iteration if the previous step is satisfied.</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genomeString)){</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">str_sub</span>(genomeString, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> nucleotide){</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>      count <span class="ot">&lt;-</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(count)</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a><span class="fu">nucleotide_frequency</span>(<span class="st">"ACTTGCGGGTATCGAG"</span>, <span class="st">"G"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<hr>
</section>
<section id="using-the-function-function-on-a-randomized-genome" class="level3">
<h3 class="anchored" data-anchor-id="using-the-function-function-on-a-randomized-genome">2.1: Using the Function Function on a Randomized Genome</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>nt_sample <span class="ot">&lt;-</span> <span class="fu">sample</span>(nt_names, <span class="at">size =</span> <span class="dv">2000</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>nt_sample <span class="ot">&lt;-</span> <span class="fu">paste</span>(nt_sample, <span class="at">collapse =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nucleotide_frequency</span>(nt_sample, <span class="st">"C"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 496</code></pre>
</div>
</div>
<ul>
<li><p>23.8% of the nucleotides in this randomized dataset are cytosines</p>
<ul>
<li>code is random, but this character</li>
</ul></li>
</ul>
</section>
<section id="generating-a-function-to-create-randomized-genome" class="level3">
<h3 class="anchored" data-anchor-id="generating-a-function-to-create-randomized-genome">2.2 Generating a function to create randomized genome</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>nt_names <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>rnd_genome <span class="ot">&lt;-</span> <span class="cf">function</span>(k){</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">215</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  rnd_string <span class="ot">&lt;-</span> <span class="st">""</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>k){</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>    rnd_string <span class="ot">&lt;-</span> <span class="fu">paste</span>(rnd_string, <span class="fu">sample</span>(nt_names,<span class="dv">1</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>), <span class="at">sep=</span><span class="st">""</span>)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(rnd_string)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rnd_genome</span>(<span class="dv">9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TGGAATCTT"</code></pre>
</div>
</div>
<section id="from-single-vector-multiple-vector-substrings" class="level4">
<h4 class="anchored" data-anchor-id="from-single-vector-multiple-vector-substrings">From single-vector multiple-vector Substrings</h4>
<ul>
<li><code>str_sub</code> or substrings, allow the function to extract characters based on a starting and and ending point, which can be valuable for extracting vectors of a specific length, or with specific values. This function will be valuable for genomic data to identify patterns and extract vectors with specific lengths.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>string_sample <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="co"># subset strings extract a value based on the </span></span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>string_sample <span class="ot">&lt;-</span> string_sample <span class="sc">%&gt;%</span> </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(nt_sample, <span class="at">start =</span> <span class="dv">1000</span>, <span class="at">end =</span> <span class="dv">1005</span>))</span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Negative vector position indicated the position to start and end from is being measured starting at the end of the dataset</span></span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="co"># When data from variables is added to a substring, they are removed and cannot be added to a new substring that overlaps it</span></span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a><span class="fu">str_sub</span>(nt_sample, <span class="at">start =</span> <span class="sc">-</span><span class="dv">1000</span>, <span class="at">end =</span> <span class="sc">-</span><span class="dv">995</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TTAATA"</code></pre>
</div>
</div>
</section>
<section id="appending-a-new-list-containing-a-substring-of-nucleotides" class="level4">
<h4 class="anchored" data-anchor-id="appending-a-new-list-containing-a-substring-of-nucleotides">Appending a new list containing a substring of nucleotides</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 'append' is useful for generating a reusable variable containing a vector that can be called back for later tests</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>string_sample <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a><span class="co"># subset strings exctract a value based on the </span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>string_sample <span class="ot">&lt;-</span> string_sample <span class="sc">%&gt;%</span> </span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(nt_sample, <span class="at">start =</span> <span class="dv">1000</span>, <span class="at">end =</span> <span class="dv">1005</span>))</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>string_sample</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CTTAAT"</code></pre>
</div>
</div>
</section>
<section id="using-multiple-vector-substrings-to-generate-paired-nucleotides-in-a-dataset" class="level4">
<h4 class="anchored" data-anchor-id="using-multiple-vector-substrings-to-generate-paired-nucleotides-in-a-dataset">Using multiple-vector substrings to generate paired nucleotides in a dataset</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>generate_2_mer <span class="ot">&lt;-</span> <span class="cf">function</span>(string_sample) {</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  list_codon <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># this function will run from 1 to the end of the supplied genome sting (-1 prevents the function from adding a variable that does not contain 2 variables)</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nchar</span>(string_sample) <span class="sc">-</span> <span class="dv">1</span>)){</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>  list_codon <span class="ot">&lt;-</span> list_codon <span class="sc">%&gt;%</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">append</span>(<span class="fu">str_sub</span>(string_sample, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> <span class="dv">1</span>))</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_codon)</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_2_mer</span>(string_sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "CT" "TT" "TA" "AA" "AT"</code></pre>
</div>
</div>
</section>
</section>
<section id="using-multiple-vector-substrings-to-generate-codons-in-a-dataset" class="level3">
<h3 class="anchored" data-anchor-id="using-multiple-vector-substrings-to-generate-codons-in-a-dataset">2.3 Using multiple-vector substrings to generate codons in a dataset</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>generate_codons <span class="ot">&lt;-</span> <span class="cf">function</span>(string_sample){</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  list_codon <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a><span class="co"># for each iteration i in the sequance 1 through all charictars in the provided string </span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="do">## -2 to prevent a partial codon</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="do">## by = 3 to make the function shift 3 rather then shifting 1 and counting 3 for each iteration</span></span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nchar</span>(string_sample) <span class="sc">-</span> <span class="dv">2</span>, <span class="at">by =</span> <span class="dv">3</span>)) {</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    list_codon <span class="ot">&lt;-</span> list_codon <span class="sc">%&gt;%</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>      <span class="do">##append adds an additional vector to the variable, without clearing what was generated previously. end = i + 2 means the function will count 3 out from the current variable being iterated, which jumps 3 each iteration. </span></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>      <span class="fu">append</span>(<span class="fu">str_sub</span>(string_sample, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> <span class="dv">2</span>))</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_codon)</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_codons</span>(<span class="fu">rnd_genome</span>(<span class="dv">200</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "TGG" "AAT" "CTT" "TAA" "TGT" "TCC" "GAC" "CTT" "TGA" "GCC" "TCA" "AGT"
[13] "TGG" "ACT" "ACC" "GCG" "GCC" "CCC" "GAC" "CGA" "GAG" "CTA" "ACT" "CTA"
[25] "AAT" "GAT" "GCA" "AGC" "AGC" "GCG" "TGC" "CGA" "GTC" "TCT" "GTG" "GAC"
[37] "AGG" "ATC" "ACA" "TTG" "GAT" "GTG" "AGC" "CAA" "GAT" "CTC" "GAG" "ATG"
[49] "AAA" "AAC" "AAG" "TAG" "GTT" "CTC" "TAT" "CGC" "TGG" "CTC" "CAA" "TCA"
[61] "TTT" "TTA" "TCC" "TTT" "GCT" "GCA"</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>generate_k_mer <span class="ot">&lt;-</span> <span class="cf">function</span>(string, <span class="at">k =</span> <span class="dv">3</span>) {</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  list_codon <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">nchar</span>(string) <span class="sc">-</span> k <span class="sc">+</span> <span class="dv">1</span>, <span class="at">by =</span> k)) {</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>    list_codon <span class="ot">&lt;-</span> list_codon <span class="sc">%&gt;%</span></span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>      <span class="fu">append</span>(<span class="fu">str_sub</span>(string, <span class="at">start =</span> i, <span class="at">end =</span> i <span class="sc">+</span> k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb46-9"><a href="#cb46-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list_codon)</span>
<span id="cb46-10"><a href="#cb46-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb46-11"><a href="#cb46-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb46-12"><a href="#cb46-12" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_k_mer</span>(<span class="fu">rnd_genome</span>(<span class="dv">9</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TGG" "AAT" "CTT"</code></pre>
</div>
</div>
<ul>
<li>This function can</li>
</ul>
</section>
<section id="using-functions-to-generate-genome-extract-strings-of-length-k-and-find-the-frequency-of-nucleotides" class="level3">
<h3 class="anchored" data-anchor-id="using-functions-to-generate-genome-extract-strings-of-length-k-and-find-the-frequency-of-nucleotides">2.4 Using functions to generate genome, extract strings of length k, and find the frequency of nucleotides</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">generate_k_mer</span>(<span class="fu">rnd_genome</span>(<span class="dv">1000</span>),<span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "TGGAAT" "CTTTAA" "TGTTCC" "GACCTT" "TGAGCC" "TCAAGT" "TGGACT" "ACCGCG"
  [9] "GCCCCC" "GACCGA" "GAGCTA" "ACTCTA" "AATGAT" "GCAAGC" "AGCGCG" "TGCCGA"
 [17] "GTCTCT" "GTGGAC" "AGGATC" "ACATTG" "GATGTG" "AGCCAA" "GATCTC" "GAGATG"
 [25] "AAAAAC" "AAGTAG" "GTTCTC" "TATCGC" "TGGCTC" "CAATCA" "TTTTTA" "TCCTTT"
 [33] "GCTGCA" "TACTTA" "TGGCCC" "TCCAAC" "CTTGCG" "AATTGC" "GGCATT" "ACGTTC"
 [41] "TATGCA" "CGAAAG" "AGGCAC" "GAGATG" "TATATC" "ACTCTT" "CAGCTA" "TCTGAC"
 [49] "ACAGGT" "AGGCAT" "TGGATG" "TCAGCG" "TACGAC" "CGGGGT" "AGGCAA" "CCCCTT"
 [57] "CTGTTG" "CCCCGC" "TGGCCG" "GATGCC" "AAGGTG" "TTTTAC" "ATCGGT" "ATTTTA"
 [65] "AAGATG" "TGAACT" "TAAGTA" "ACCTAC" "TTAGCT" "GACGGG" "AAGGTG" "CACATG"
 [73] "TATGTG" "TGACTC" "TACACC" "AAGATG" "CACTTA" "GGCATC" "AATAAA" "AGTTGC"
 [81] "CGGTTT" "GTAATC" "CTTGAC" "AGTAAT" "CGAGAT" "AATTAC" "TTGCGG" "GCACAT"
 [89] "AACCTA" "CTCGGT" "TCGTCC" "CCCTTA" "GCGTAC" "GGGGGG" "GTGGGG" "AACCAT"
 [97] "CAGCGT" "TCGTAT" "GTTAGT" "CCTCCG" "GCATTT" "TTGGTC" "ACGGCC" "CTCCAT"
[105] "GAAATA" "CATACT" "AGCGTG" "CCATTC" "CGGTCA" "AATGGC" "CCTGCG" "AAGGAT"
[113] "GCGGTG" "CGGTAA" "GCGTGT" "GGGCCA" "TACTTG" "GGCAGA" "TTGAGT" "TATAAG"
[121] "AAACAG" "GCAAAT" "TGGGAA" "CTAGTG" "CGGCAC" "ATGTAC" "AGCTCG" "CACTTC"
[129] "CTGGGG" "GGCGAC" "TACCGA" "ATTCCT" "AGTATG" "TCTGAG" "TAGCGT" "ACCGGG"
[137] "CCAGCT" "TTGGGT" "CCTCTT" "ACCTTG" "TAATGG" "GATCGG" "CCTTAT" "GGGATG"
[145] "CCATGA" "CGACAC" "TGCGGG" "AGGTTA" "ATAAGT" "GTCCGC" "GATAAA" "GCCTCG"
[153] "TGGTCG" "AAATCT" "GTTACG" "TGCTCT" "TCTGTC" "ATTTGC" "CATATT" "GGTAAA"
[161] "GTTCAA" "GAGCCA" "CCCTAT" "CTATTG" "ACTCCT" "AGATCT"</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nucleotide_frequency</span>(<span class="fu">rnd_genome</span>(<span class="dv">1000</span>),<span class="st">"G"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 257</code></pre>
</div>
</div>
<ul>
<li>putting together everything completed in this chapter, the written code can now perform genomic analysis with pre-written functions that are proven to function as intended, and can be utilized for any genome.</li>
</ul>
</section>
<section id="generating-a-function-to-find-the-frequency-of-a-specific-nucleotide-string" class="level3">
<h3 class="anchored" data-anchor-id="generating-a-function-to-find-the-frequency-of-a-specific-nucleotide-string">2.5 Generating a function to find the frequency of a specific nucleotide string</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>nt_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(string, pattern) {</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  nt_matches <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  len_string <span class="ot">&lt;-</span> <span class="fu">nchar</span>(string)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  len_pattern <span class="ot">&lt;-</span> <span class="fu">str_length</span>(pattern)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(len_string <span class="sc">-</span> len_pattern <span class="sc">+</span> <span class="dv">1</span>)){</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(string, i, i <span class="sc">+</span> len_pattern <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">==</span> pattern){</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>      nt_matches <span class="ot">=</span> nt_matches <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(nt_matches)</span>
<span id="cb52-12"><a href="#cb52-12" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb52-13"><a href="#cb52-13" aria-hidden="true" tabindex="-1"></a><span class="fu">nt_pattern</span>(<span class="fu">rnd_genome</span>(<span class="dv">2000</span>), <span class="st">"GACCTT"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>rosalind_string <span class="ot">&lt;-</span> <span class="st">"TTAGTCCCCAGTCCCCAGTCCCCTCCAGTCCCCGCAGTCCCCCAGTCCCCAGTCCCCCAGTCCCCAGTCCCCAGTCCCCAGCCAGTCCCCACCGGTGTGGTAGTCCCCCAAGTCCCCAAGTCCCCAGTGATAACAGTCCCCTTCTCTAAGTCCCCAGTCCCCGAGTCCCCAGTTGAGTCCCCCTAGTCCCCGCCTATAGTCCCCCCACGAGTCCCCTGAAGTCCCCTGAAAGTCCCCCTGACGCAAGTCCCCTAGTCCCCCAGTCCCCAGAAGTCCCCCAGTCCCCTAAGTCCCCTAGAGTCCCCAGTCCCCGAGAGTCCCCTGTAAGTCCCCCTCAGTCCCCGGCTCGAGTCCCCGATGAGTCCCCGAGTCCCCCCGAGTCCCCGGTTAGTCCCCAAGTCCCCGAGTCCCCGAGTCCCCTGAAGTCCCCGAGTCCCCTCGAGTCCCCAGTCCCCGCTAGTCCCCCTTAGTCCCCAGTCCCCGAGTCCCCAGTCCCCAGTCCCCAAGTCCCCCGTGGAGAAGTCCCCGCAGTCCCCAGTCCCCTCGATTAGTCCCCATGCGATAGTCCCCCAGTCCCCTGAGTCCCCAGTCCCCAAGTCCCCGTTAGTCCCCGAGTCCCCAAAATTAGTCCCCGAAGTCCCCCCGTAGTCCCCTGTGAGTCCCCGAGTCCCCAGTCCCCTTACGAGTCCCCGTCCAGTCCCCTGATTATATGAGAGTCCCCTTGGAGTCCCCTAGTCCCCTAAGTCCCCAGAGTGATTCTTAGTCCCCAAGTCCCCAGTCCCCGCTAGTCCCCATAGTCCCCAGTCCCCCGCAGTCCCCCTACCTCAGTCCCCAAGTCCCCCAGTCCCCCAGTCCCCGGTATTAGTCCCCGAAGTCCCCGAGTCCCCATACTCAAGTCCCCCAGTCCCCGGATGGTAGAGTCCCCAGTCCCCTAGTCCCCCCGAGTCCCCAGTCCCCCAGTCCCCACGGCGGCTAAAGTCCCCTAGTCCCCAGTCCCCGATGCAGTCCCCAAGTCCCCAGTCCCC"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nt_pattern</span>(rosalind_string, <span class="st">"AGTCCCCAG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 25</code></pre>
</div>
</div>
<ul>
<li>The code suggests this pattern appeared 25 times, which was the correct answer, meaning the function worked as intended!</li>
</ul>
<hr>
</section>
</section>
</section>
<section id="chapter-3-bioinformatic-tools-pt.-3" class="level1">
<h1>Chapter 3: Bioinformatic Tools pt.&nbsp;3</h1>
<section id="using-functions-to-analyze-patterns-in-a-genome-sequence" class="level2">
<h2 class="anchored" data-anchor-id="using-functions-to-analyze-patterns-in-a-genome-sequence">Using Functions to analyze patterns in a genome sequence</h2>
<ul>
<li>The previous steps have been building the tools to start making more complex functions capable of inputting a genome and looking for repeating code, in specific places.</li>
</ul>
<section id="combining-generate_k_mer-and-nt_pattern-to-build-a-function-generate_frequent_k_mer" class="level3">
<h3 class="anchored" data-anchor-id="combining-generate_k_mer-and-nt_pattern-to-build-a-function-generate_frequent_k_mer">3.1 Combining <code>generate_k_mer</code> and <code>nt_pattern</code> to build a function <code>generate_frequent_k_mer</code></h3>
<ul>
<li><p>This step will require putting together the concepts building up to and including generating strings of a specific length, and recognizing patterns of a specific length</p>
<ul>
<li>The goal of this step is to generate a function that will take a <code>genome string</code> and find repeats of a function that are a specific length <code>k</code>, outputting a list using the <code>unique()</code> function to remove repeats</li>
</ul></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>generate_frequent_k_mer <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, <span class="at">k=</span><span class="dv">9</span>) {</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># generate all k-mers from the genome by the length k using `generate_k_mer` function previously created</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  strings <span class="ot">&lt;-</span> <span class="fu">generate_k_mer</span>(genome, k)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Run strings through the unique function to generate a set where each unique string length `k` is not repeated. This allows us to compile a string </span></span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  strings <span class="ot">&lt;-</span> <span class="fu">unique</span>(strings)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># count the frequency of each k-mer using the `nt_patterns` function previously created</span></span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  k_mer_count <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>,<span class="fu">length</span>(strings))</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(strings)) {</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genome)<span class="sc">-</span>(k<span class="sc">+</span><span class="dv">1</span>)) {</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (strings[i]<span class="sc">==</span> <span class="fu">str_sub</span>(genome,j,j<span class="sc">+</span>(k<span class="dv">-1</span>))){</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>      k_mer_count[i]<span class="ot">&lt;-</span>k_mer_count[i] <span class="sc">+</span><span class="dv">1</span></span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># order the k-mers by frequency and return the top k-mers and their counts as a list</span></span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a>  max_count <span class="ot">&lt;-</span> <span class="fu">max</span>(k_mer_count)</span>
<span id="cb57-19"><a href="#cb57-19" aria-hidden="true" tabindex="-1"></a>  ordered_k_mers <span class="ot">&lt;-</span> strings[k_mer_count <span class="sc">==</span> max_count]</span>
<span id="cb57-20"><a href="#cb57-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">noquote</span>(ordered_k_mers))</span>
<span id="cb57-21"><a href="#cb57-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb57-22"><a href="#cb57-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-23"><a href="#cb57-23" aria-hidden="true" tabindex="-1"></a> <span class="fu">generate_frequent_k_mer</span>(<span class="st">"ACGTTGCATGTCGCATGATGCATGAGAGCT</span></span>
<span id="cb57-24"><a href="#cb57-24" aria-hidden="true" tabindex="-1"></a><span class="st">"</span>,<span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] GCAT CATG</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co">#generate_frequent_k_mer("GGGCCGGGTACATCGTAAGCCGTATGGGCCGGGTACATCGTAAGGACGATCGGACGATCGGGCCGGGTACATCGTAAGGGCCGGGTGGGCCGGGTGGACGATCACATCGTAAGGGCCGGGTACATCGTAAGCCGTATACATCGTAAGCCGTATGGACGATCGGGCCGGGTTCCTAGGGGGCCGGGTGGACGATCGCCGTATGGACGATCACATCGTAAACATCGTAAGGACGATCGGGCCGGGTGGGCCGGGTTCCTAGGTCCTAGGACATCGTAAGCCGTATACATCGTAAGGACGATCGGACGATCACATCGTAAGGGCCGGGTTCCTAGGGGGCCGGGTGCCGTATGGGCCGGGTGGACGATCGCCGTATGGGCCGGGTACATCGTAAGCCGTATGGACGATCGCCGTATGCCGTATGGGCCGGGTGGACGATCGGGCCGGGTGGACGATCGGACGATCGGGCCGGGTGGACGATCGCCGTATTCCTAGGGGACGATCACATCGTAATCCTAGGGGACGATCACATCGTAATCCTAGGGGGCCGGGTTCCTAGGGCCGTATTCCTAGGGGACGATCGGGCCGGGTGGGCCGGGTACATCGTAAGGACGATCACATCGTAAGGACGATCGGGCCGGGTGGGCCGGGTGGACGATCGCCGTATGCCGTATGGACGATCGGGCCGGGTTCCTAGGGGACGATCGCCGTATGGGCCGGGTTCCTAGGGGACGATCGGGCCGGGTACATCGTAAGGGCCGGGTACATCGTAATCCTAGGTCCTAGGGGACGATCGGACGATCACATCGTAAACATCGTAAGGACGATC",11)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="generating-the-complementary-dna-strand" class="level3">
<h3 class="anchored" data-anchor-id="generating-the-complementary-dna-strand">3.2 Generating the complementary DNA Strand</h3>
<ul>
<li><p>as of now this code only generates the genome of one strand, The complementary stand of DNA reads in the <strong>reverse direction</strong>.</p>
<ul>
<li>Because the complement strand uses the same</li>
</ul></li>
</ul>
</section>
<section id="utilizing-rev-to-generate-complement-dna-strand" class="level3">
<h3 class="anchored" data-anchor-id="utilizing-rev-to-generate-complement-dna-strand">Utilizing <code>rev()</code> to generate complement DNA strand</h3>
<ul>
<li><p>First step toward generating a complementary genome is to be able to write it in the correction orientation. The complementary strands are <strong>antiparallel</strong>, meaning they run in opposite directions. The 5’ end of one strand is paired with the 3’ end of the other strand. In contrast, the reading strand runs in the 5’ to 3’ direction, which is the direction in which DNA is read and transcribed into RNA.<br>
</p>
<hr></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>genome <span class="ot">&lt;-</span> <span class="fu">rnd_genome</span>(<span class="dv">10</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rev</span>(<span class="fu">c</span>(<span class="st">"C"</span>,<span class="st">"T"</span>,<span class="st">"G"</span>,<span class="st">"A"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "G" "T" "C"</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span> (genome)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TGGAATCTTT"</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Gemerate a vector splitting each character into a vector</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="do">## vectors are required for the `rev()` function, which does not take string inputs</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="do">## [[1]] stops the function from returning a </span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>char_vector <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(<span class="at">x =</span> <span class="st">'TTTCTAAGGT'</span>, <span class="at">split =</span> <span class="st">""</span> )[[<span class="dv">1</span>]]</span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Reverse the character vector</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>rev_char_vector <span class="ot">&lt;-</span> <span class="fu">rev</span>(<span class="at">x =</span> char_vector)</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Collapse the reversed character vector back into a string, this puts it back in the format that it was originally displayed</span></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>reversed_sequence <span class="ot">&lt;-</span> <span class="fu">paste</span>(rev_char_vector, <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Print the reversed sequence</span></span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(reversed_sequence)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TGGAATCTTT"</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>reverse_complement <span class="ot">&lt;-</span> <span class="cf">function</span>(GenomeSubString) {</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a>  complement <span class="ot">&lt;-</span> <span class="st">""</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop is used to write the complementary base for each nucleotide in the given sequence: for each individual vector, the loop will run through each nucleotide and write the complement in its place. </span></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>  <span class="do">## if/else means the genome is only scanned through once, rather then for each nucleotide.</span></span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Paste0 is used instead of paste to combine with no seperator " " which is added in the normal paste function</span></span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (nucleotide <span class="cf">in</span> <span class="fu">strsplit</span>(GenomeSubString, <span class="st">""</span>)[[<span class="dv">1</span>]]) {</span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (nucleotide <span class="sc">==</span> <span class="st">"A"</span>) {</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a>      complement <span class="ot">&lt;-</span> <span class="fu">paste0</span>(complement, <span class="st">"T"</span>)</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> <span class="cf">if</span> (nucleotide <span class="sc">==</span> <span class="st">"T"</span>) {</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>      complement <span class="ot">&lt;-</span> <span class="fu">paste0</span>(complement, <span class="st">"A"</span>)</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> <span class="cf">if</span> (nucleotide <span class="sc">==</span> <span class="st">"C"</span>) {</span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>      complement <span class="ot">&lt;-</span> <span class="fu">paste0</span>(complement, <span class="st">"G"</span>)</span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> <span class="cf">if</span> (nucleotide <span class="sc">==</span> <span class="st">"G"</span>) {</span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>      complement <span class="ot">&lt;-</span> <span class="fu">paste0</span>(complement, <span class="st">"C"</span>)</span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># After rewriting the genome with the complement stand, rev() is used to flip it to reading in the correct direction (5` to 3`) used in DNA transcription. </span></span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Becuse this function needs vectors and not character string, the genome is split into a substring for each value to be reversed and then collapsed back into a character string. </span></span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>  reverse_complement <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">rev</span>(<span class="fu">strsplit</span>(complement, <span class="st">""</span>)[[<span class="dv">1</span>]]), <span class="at">collapse =</span> <span class="st">""</span>)</span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(reverse_complement)</span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co">#data &lt;- scan('C:/School/23SPDAY/Biostatistics/Rever_string_rosalind.txt', </span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>             <span class="co">#what = character())</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co">#genomeString &lt;- data[1]</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-7"><a href="#cb67-7" aria-hidden="true" tabindex="-1"></a><span class="co">#reverse_complement(genomeString)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
</section>
<section id="chapter-4-from-groups-to-sequences-finding-patterns-in-the-genome" class="level1">
<h1>Chapter 4: From Groups to Sequences – Finding patterns in the genome</h1>
<ul>
<li>Up to this point the goal has been to build small functions that can parse through the genome and pull out relevant information. This chapter goes a step further toward using these functions to uncover the location of dnaA boxes in a bacterial genome.</li>
</ul>
<section id="clump-finding" class="level2">
<h2 class="anchored" data-anchor-id="clump-finding">4.1 Clump Finding –</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>initialize_k_mer_dict <span class="ot">&lt;-</span> <span class="cf">function</span>(k) {</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>  k_mers_dict <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">rep</span>(<span class="fu">list</span>(nucleotides), k)) <span class="sc">%&gt;%</span></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unite</span>(<span class="st">"k_mers"</span>, <span class="fu">everything</span>(), <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">count =</span> <span class="dv">0</span>)</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(k_mers_dict)</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="unoptimized-version-of-clump_funding-function" class="level3">
<h3 class="anchored" data-anchor-id="unoptimized-version-of-clump_funding-function">Unoptimized version of clump_funding Function</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>initialize_k_mer_dict <span class="ot">&lt;-</span> <span class="cf">function</span>(k) {</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"C"</span>, <span class="st">"G"</span>, <span class="st">"T"</span>)</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a>  k_mers_dict <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">rep</span>(<span class="fu">list</span>(nucleotides), k)) <span class="sc">%&gt;%</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unite</span>(<span class="st">"k_mers"</span>, <span class="fu">everything</span>(), <span class="at">remove =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>() <span class="sc">%&gt;%</span></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mutate</span>(<span class="at">count =</span> <span class="dv">0</span>)</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(k_mers_dict)</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genomeString, L, t, k) {</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>  genome_length <span class="ot">&lt;-</span> <span class="fu">nchar</span>(genomeString)</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>  frequent_k_mers <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(genome_length <span class="sc">-</span> L <span class="sc">+</span> <span class="dv">1</span>)) {</span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a>    window <span class="ot">&lt;-</span> <span class="fu">substr</span>(genomeString, j, j <span class="sc">+</span> L <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a>    k_mers_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(k)</span>
<span id="cb69-19"><a href="#cb69-19" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb69-20"><a href="#cb69-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(L <span class="sc">-</span> k <span class="sc">+</span> <span class="dv">1</span>)) {</span>
<span id="cb69-21"><a href="#cb69-21" aria-hidden="true" tabindex="-1"></a>      k_mer <span class="ot">&lt;-</span> <span class="fu">substr</span>(window, i, i <span class="sc">+</span> k <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb69-22"><a href="#cb69-22" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb69-23"><a href="#cb69-23" aria-hidden="true" tabindex="-1"></a>      k_mers_dict[k_mers_dict<span class="sc">$</span>k_mers <span class="sc">==</span> k_mer, <span class="st">"count"</span>] <span class="ot">&lt;-</span> k_mers_dict[k_mers_dict<span class="sc">$</span>k_mers <span class="sc">==</span> k_mer, <span class="st">"count"</span>] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb69-24"><a href="#cb69-24" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb69-25"><a href="#cb69-25" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb69-26"><a href="#cb69-26" aria-hidden="true" tabindex="-1"></a>    k_mers_dict_filtered <span class="ot">&lt;-</span> k_mers_dict <span class="sc">%&gt;%</span></span>
<span id="cb69-27"><a href="#cb69-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">filter</span>(count <span class="sc">&gt;=</span> t) <span class="sc">%&gt;%</span></span>
<span id="cb69-28"><a href="#cb69-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">pull</span>(k_mers)</span>
<span id="cb69-29"><a href="#cb69-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb69-30"><a href="#cb69-30" aria-hidden="true" tabindex="-1"></a>    frequent_k_mers <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(frequent_k_mers, k_mers_dict_filtered))</span>
<span id="cb69-31"><a href="#cb69-31" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb69-32"><a href="#cb69-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb69-33"><a href="#cb69-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(frequent_k_mers)</span>
<span id="cb69-34"><a href="#cb69-34" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="optimized-version-1" class="level3">
<h3 class="anchored" data-anchor-id="optimized-version-1">Optimized Version 1</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The clump_finding function takes in a genome (a string representing a DNA sequence),</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="co"># L (the length of the region in the genome to consider), t (the threshold of occurrences</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a><span class="co"># for a k-mer to be considered frequent), and k (the length of the k-mer to find).</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, L, t, k){</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initializes a dictionary of all possible k-mers of length k.</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  kmer_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(k)</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Retrieves the list of all possible k-mers from the dictionary.</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>  kmer_list <span class="ot">&lt;-</span> kmer_dict<span class="sc">$</span>k_mers</span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Counts the number of unique k-mers (k-mer variations) in the list.</span></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>  nk <span class="ot">&lt;-</span> <span class="fu">length</span>(kmer_list)</span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initializes a list to store counts of each k-mer's occurrences, setting all counts to 0 initially.</span></span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>  counts_list <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, nk)</span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Counts the total number of characters in the genome.</span></span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a>  ng <span class="ot">&lt;-</span> <span class="fu">nchar</span>(genome)</span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Extracts the first L characters from the genome.</span></span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a>  first_L <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, <span class="dv">1</span>, L)</span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Counts the occurrences of each k-mer in the first L characters of the genome.</span></span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nk){</span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a>    current_pattern <span class="ot">&lt;-</span> kmer_list[i]</span>
<span id="cb70-23"><a href="#cb70-23" aria-hidden="true" tabindex="-1"></a>    counts_list[i] <span class="ot">&lt;-</span> <span class="fu">nt_pattern</span>(first_L, current_pattern)</span>
<span id="cb70-24"><a href="#cb70-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb70-25"><a href="#cb70-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-26"><a href="#cb70-26" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Finds all k-mers that occur at least t times in the first L characters of the genome.</span></span>
<span id="cb70-27"><a href="#cb70-27" aria-hidden="true" tabindex="-1"></a>  freq_kmers <span class="ot">&lt;-</span> kmer_list[counts_list <span class="sc">&gt;=</span> t]</span>
<span id="cb70-28"><a href="#cb70-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-29"><a href="#cb70-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Slides a window of length L across the genome, one character at a time.</span></span>
<span id="cb70-30"><a href="#cb70-30" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>(ng <span class="sc">-</span> (L <span class="sc">-</span> <span class="dv">2</span>))){</span>
<span id="cb70-31"><a href="#cb70-31" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Identifies the k-mer that will be removed from the window.</span></span>
<span id="cb70-32"><a href="#cb70-32" aria-hidden="true" tabindex="-1"></a>    rem_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">-</span> <span class="dv">1</span>, j <span class="sc">-</span> <span class="dv">1</span> <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb70-33"><a href="#cb70-33" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Identifies the k-mer that will be added to the window.</span></span>
<span id="cb70-34"><a href="#cb70-34" aria-hidden="true" tabindex="-1"></a>    add_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, L <span class="sc">+</span> j<span class="dv">-1</span> <span class="sc">-</span> (k<span class="dv">-1</span>),L <span class="sc">+</span> j<span class="dv">-1</span>)</span>
<span id="cb70-35"><a href="#cb70-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-36"><a href="#cb70-36" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Decreases the count of the k-mer being removed from the window by 1.</span></span>
<span id="cb70-37"><a href="#cb70-37" aria-hidden="true" tabindex="-1"></a>    counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> rem_pattern)] <span class="ot">&lt;-</span> counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> rem_pattern)] <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb70-38"><a href="#cb70-38" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Increases the count of the k-mer being added to the window by 1.</span></span>
<span id="cb70-39"><a href="#cb70-39" aria-hidden="true" tabindex="-1"></a>    counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> add_pattern)] <span class="ot">&lt;-</span> counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> add_pattern)] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb70-40"><a href="#cb70-40" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb70-41"><a href="#cb70-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># If the count of the added k-mer reaches the threshold t, it is added to the list of frequent k-mers.</span></span>
<span id="cb70-42"><a href="#cb70-42" aria-hidden="true" tabindex="-1"></a>  add_pattern_index <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_list <span class="sc">==</span> add_pattern)</span>
<span id="cb70-43"><a href="#cb70-43" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(add_pattern_index) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> counts_list[add_pattern_index] <span class="sc">&gt;=</span> t) {</span>
<span id="cb70-44"><a href="#cb70-44" aria-hidden="true" tabindex="-1"></a>      freq_kmers <span class="ot">&lt;-</span> <span class="fu">append</span>(freq_kmers, add_pattern)</span>
<span id="cb70-45"><a href="#cb70-45" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb70-46"><a href="#cb70-46" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb70-47"><a href="#cb70-47" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb70-48"><a href="#cb70-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">unique</span>(freq_kmers))</span>
<span id="cb70-49"><a href="#cb70-49" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<ul>
<li>Provided by the professor, but still not working correctly…</li>
</ul>
</section>
<section id="optimized-version-3" class="level3">
<h3 class="anchored" data-anchor-id="optimized-version-3">Optimized Version 3</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, L, t, k){</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  kmer_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(k)</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#kmer_list &lt;- kmer_dict$k_mers</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#nk &lt;- length(kmer_list)</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#counts_list &lt;- rep(0, nk)</span></span>
<span id="cb71-6"><a href="#cb71-6" aria-hidden="true" tabindex="-1"></a>  ng <span class="ot">&lt;-</span> <span class="fu">nchar</span>(genome)</span>
<span id="cb71-7"><a href="#cb71-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-8"><a href="#cb71-8" aria-hidden="true" tabindex="-1"></a>  initial_window <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, <span class="dv">1</span>, L)</span>
<span id="cb71-9"><a href="#cb71-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-10"><a href="#cb71-10" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(L <span class="sc">-</span> (k <span class="sc">-</span> <span class="dv">1</span>))){</span>
<span id="cb71-11"><a href="#cb71-11" aria-hidden="true" tabindex="-1"></a>    curr_kmer <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(initial_window, i, i <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb71-12"><a href="#cb71-12" aria-hidden="true" tabindex="-1"></a>    curr_count <span class="ot">&lt;-</span> <span class="fu">nt_pattern</span>(initial_window, curr_kmer)</span>
<span id="cb71-13"><a href="#cb71-13" aria-hidden="true" tabindex="-1"></a>    curr_row <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_dict<span class="sc">$</span>k_mers <span class="sc">==</span> curr_kmer)</span>
<span id="cb71-14"><a href="#cb71-14" aria-hidden="true" tabindex="-1"></a>    kmer_dict<span class="sc">$</span>count[curr_row] <span class="ot">&lt;-</span> curr_count</span>
<span id="cb71-15"><a href="#cb71-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-16"><a href="#cb71-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-17"><a href="#cb71-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"I'm done searching the initial Window!"</span>)</span>
<span id="cb71-18"><a href="#cb71-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-19"><a href="#cb71-19" aria-hidden="true" tabindex="-1"></a>  candidates <span class="ot">&lt;-</span> kmer_dict <span class="sc">%&gt;%</span></span>
<span id="cb71-20"><a href="#cb71-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">filter</span>(count <span class="sc">&gt;=</span> t) <span class="sc">%&gt;%</span></span>
<span id="cb71-21"><a href="#cb71-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pull</span>(k_mers)</span>
<span id="cb71-22"><a href="#cb71-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-23"><a href="#cb71-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"I found initial frequent kmers: "</span>, candidates))</span>
<span id="cb71-24"><a href="#cb71-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Now I'm sliding the window through the rest of the genome!"</span>)</span>
<span id="cb71-25"><a href="#cb71-25" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb71-26"><a href="#cb71-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>(ng <span class="sc">-</span> (L <span class="sc">-</span> <span class="dv">2</span>))){</span>
<span id="cb71-27"><a href="#cb71-27" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>((j <span class="sc">%%</span> <span class="dv">1000</span>) <span class="sc">==</span> <span class="dv">0</span>){</span>
<span id="cb71-28"><a href="#cb71-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Hey there! The left edge of my window is currently at nucleotide "</span>, j))</span>
<span id="cb71-29"><a href="#cb71-29" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-30"><a href="#cb71-30" aria-hidden="true" tabindex="-1"></a>    rem_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">-</span> <span class="dv">1</span>, j <span class="sc">-</span> <span class="dv">1</span> <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb71-31"><a href="#cb71-31" aria-hidden="true" tabindex="-1"></a>    add_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">+</span> (L <span class="sc">-</span> <span class="dv">2</span>) <span class="sc">-</span> (k <span class="sc">-</span> <span class="dv">1</span>), j <span class="sc">+</span> (L <span class="sc">-</span> <span class="dv">2</span>))</span>
<span id="cb71-32"><a href="#cb71-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-33"><a href="#cb71-33" aria-hidden="true" tabindex="-1"></a>    rem_row <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_dict<span class="sc">$</span>k_mers <span class="sc">==</span> rem_pattern)</span>
<span id="cb71-34"><a href="#cb71-34" aria-hidden="true" tabindex="-1"></a>    add_row <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_dict<span class="sc">$</span>k_mers <span class="sc">==</span> add_pattern)</span>
<span id="cb71-35"><a href="#cb71-35" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-36"><a href="#cb71-36" aria-hidden="true" tabindex="-1"></a>    kmer_dict<span class="sc">$</span>count[rem_row] <span class="ot">&lt;-</span> kmer_dict<span class="sc">$</span>count[rem_row] <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb71-37"><a href="#cb71-37" aria-hidden="true" tabindex="-1"></a>    kmer_dict<span class="sc">$</span>count[add_row] <span class="ot">&lt;-</span> kmer_dict<span class="sc">$</span>count[add_row] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb71-38"><a href="#cb71-38" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb71-39"><a href="#cb71-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(kmer_dict<span class="sc">$</span>count[add_row] <span class="sc">&gt;=</span> t){</span>
<span id="cb71-40"><a href="#cb71-40" aria-hidden="true" tabindex="-1"></a>      candidates <span class="ot">&lt;-</span> <span class="fu">append</span>(candidates, add_pattern)</span>
<span id="cb71-41"><a href="#cb71-41" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb71-42"><a href="#cb71-42" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb71-43"><a href="#cb71-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">unique</span>(candidates))</span>
<span id="cb71-44"><a href="#cb71-44" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="optimized-version-4" class="level3">
<h3 class="anchored" data-anchor-id="optimized-version-4">Optimized Version 4</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>nt_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(text, pattern, nt_table) {</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>  pattern_length <span class="ot">&lt;-</span> <span class="fu">nchar</span>(pattern)</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>  text_length <span class="ot">&lt;-</span> <span class="fu">nchar</span>(text)</span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a>  count <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a>  pattern_int <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(pattern, <span class="st">""</span>), <span class="cf">function</span>(x) nt_table[x])</span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(text_length <span class="sc">-</span> pattern_length <span class="sc">+</span> <span class="dv">1</span>)) {</span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a>    text_sub <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(<span class="fu">substr</span>(text, i, i <span class="sc">+</span> pattern_length <span class="sc">-</span> <span class="dv">1</span>), <span class="st">""</span>), <span class="cf">function</span>(x) nt_table[x])</span>
<span id="cb72-14"><a href="#cb72-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">all</span>(text_sub <span class="sc">==</span> pattern_int)) {</span>
<span id="cb72-15"><a href="#cb72-15" aria-hidden="true" tabindex="-1"></a>      count <span class="ot">&lt;-</span> count <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb72-16"><a href="#cb72-16" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb72-17"><a href="#cb72-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb72-18"><a href="#cb72-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-19"><a href="#cb72-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(count)</span>
<span id="cb72-20"><a href="#cb72-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb72-21"><a href="#cb72-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-22"><a href="#cb72-22" aria-hidden="true" tabindex="-1"></a>initialize_k_mer_dict <span class="ot">&lt;-</span> <span class="cf">function</span>(k, nt_table) {</span>
<span id="cb72-23"><a href="#cb72-23" aria-hidden="true" tabindex="-1"></a>  nucleotides <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)</span>
<span id="cb72-24"><a href="#cb72-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-25"><a href="#cb72-25" aria-hidden="true" tabindex="-1"></a>  k_mers_int <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="fu">rep</span>(<span class="fu">list</span>(nucleotides), k)) <span class="sc">%&gt;%</span></span>
<span id="cb72-26"><a href="#cb72-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">apply</span>(<span class="dv">1</span>, paste0, <span class="at">collapse =</span> <span class="st">""</span>) <span class="sc">%&gt;%</span></span>
<span id="cb72-27"><a href="#cb72-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">unique</span>()</span>
<span id="cb72-28"><a href="#cb72-28" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-29"><a href="#cb72-29" aria-hidden="true" tabindex="-1"></a>  k_mers_char <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(k_mers_int, <span class="st">""</span>), <span class="cf">function</span>(x) <span class="fu">names</span>(nt_table)[<span class="fu">as.numeric</span>(x)])</span>
<span id="cb72-30"><a href="#cb72-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-31"><a href="#cb72-31" aria-hidden="true" tabindex="-1"></a>  k_mers_dict <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">k_mers =</span> k_mers_char, <span class="at">count =</span> <span class="dv">0</span>)</span>
<span id="cb72-32"><a href="#cb72-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-33"><a href="#cb72-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(k_mers_dict)</span>
<span id="cb72-34"><a href="#cb72-34" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb72-35"><a href="#cb72-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-36"><a href="#cb72-36" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, L, t, k) {</span>
<span id="cb72-37"><a href="#cb72-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-38"><a href="#cb72-38" aria-hidden="true" tabindex="-1"></a>  nt_table <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"C"</span> <span class="ot">=</span> <span class="dv">2</span>, <span class="st">"G"</span> <span class="ot">=</span> <span class="dv">3</span>, <span class="st">"T"</span> <span class="ot">=</span> <span class="dv">4</span>)</span>
<span id="cb72-39"><a href="#cb72-39" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-40"><a href="#cb72-40" aria-hidden="true" tabindex="-1"></a>  kmer_dict <span class="ot">&lt;-</span> <span class="fu">initialize_k_mer_dict</span>(k, nt_table)</span>
<span id="cb72-41"><a href="#cb72-41" aria-hidden="true" tabindex="-1"></a>  kmer_list <span class="ot">&lt;-</span> kmer_dict<span class="sc">$</span>k_mers</span>
<span id="cb72-42"><a href="#cb72-42" aria-hidden="true" tabindex="-1"></a>  nk <span class="ot">&lt;-</span> <span class="fu">length</span>(kmer_list)</span>
<span id="cb72-43"><a href="#cb72-43" aria-hidden="true" tabindex="-1"></a>  counts_list <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">0</span>, nk)</span>
<span id="cb72-44"><a href="#cb72-44" aria-hidden="true" tabindex="-1"></a>  ng <span class="ot">&lt;-</span> <span class="fu">nchar</span>(genome)</span>
<span id="cb72-45"><a href="#cb72-45" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-46"><a href="#cb72-46" aria-hidden="true" tabindex="-1"></a>  first_L <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, <span class="dv">1</span>, L)</span>
<span id="cb72-47"><a href="#cb72-47" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-48"><a href="#cb72-48" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nk) {</span>
<span id="cb72-49"><a href="#cb72-49" aria-hidden="true" tabindex="-1"></a>    current_pattern <span class="ot">&lt;-</span> kmer_list[i]</span>
<span id="cb72-50"><a href="#cb72-50" aria-hidden="true" tabindex="-1"></a>    counts_list[i] <span class="ot">&lt;-</span> <span class="fu">nt_pattern</span>(first_L, current_pattern, nt_table)</span>
<span id="cb72-51"><a href="#cb72-51" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb72-52"><a href="#cb72-52" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-53"><a href="#cb72-53" aria-hidden="true" tabindex="-1"></a>  freq_kmers <span class="ot">&lt;-</span> kmer_list[counts_list <span class="sc">&gt;=</span> t]</span>
<span id="cb72-54"><a href="#cb72-54" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-55"><a href="#cb72-55" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>(ng <span class="sc">-</span> L <span class="sc">+</span> <span class="dv">1</span>)) {</span>
<span id="cb72-56"><a href="#cb72-56" aria-hidden="true" tabindex="-1"></a>    rem_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, j <span class="sc">-</span> <span class="dv">1</span>, j <span class="sc">-</span> <span class="dv">1</span> <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb72-57"><a href="#cb72-57" aria-hidden="true" tabindex="-1"></a>    add_pattern <span class="ot">&lt;-</span> <span class="fu">str_sub</span>(genome, L <span class="sc">+</span> j <span class="sc">-</span> <span class="dv">1</span> <span class="sc">-</span> (k <span class="sc">-</span> <span class="dv">1</span>), L <span class="sc">+</span> j <span class="sc">-</span> <span class="dv">1</span>)</span>
<span id="cb72-58"><a href="#cb72-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-59"><a href="#cb72-59" aria-hidden="true" tabindex="-1"></a>    counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> rem_pattern)] <span class="ot">&lt;-</span> counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> rem_pattern)] <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb72-60"><a href="#cb72-60" aria-hidden="true" tabindex="-1"></a>    counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> add_pattern)] <span class="ot">&lt;-</span> counts_list[<span class="fu">which</span>(kmer_list <span class="sc">==</span> add_pattern)] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb72-61"><a href="#cb72-61" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb72-62"><a href="#cb72-62" aria-hidden="true" tabindex="-1"></a>    add_pattern_index <span class="ot">&lt;-</span> <span class="fu">which</span>(kmer_list <span class="sc">==</span> add_pattern)</span>
<span id="cb72-63"><a href="#cb72-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">length</span>(add_pattern_index) <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;&amp;</span> counts_list[add_pattern_index] <span class="sc">&gt;=</span> t) {</span>
<span id="cb72-64"><a href="#cb72-64" aria-hidden="true" tabindex="-1"></a>      freq_kmers <span class="ot">&lt;-</span> <span class="fu">append</span>(freq_kmers, add_pattern)</span>
<span id="cb72-65"><a href="#cb72-65" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb72-66"><a href="#cb72-66" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb72-67"><a href="#cb72-67" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-68"><a href="#cb72-68" aria-hidden="true" tabindex="-1"></a>  freq_kmers <span class="ot">&lt;-</span> <span class="fu">unique</span>(freq_kmers)</span>
<span id="cb72-69"><a href="#cb72-69" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-70"><a href="#cb72-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(freq_kmers)</span>
<span id="cb72-71"><a href="#cb72-71" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="code-to-test-this-function" class="level3">
<h3 class="anchored" data-anchor-id="code-to-test-this-function">4.4 Code to test this function</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>genomeString <span class="ot">&lt;-</span> <span class="st">"AAAACCCCCCAAAA"</span> </span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a><span class="fu">clump_finding</span>(genomeString, <span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A"</code></pre>
</div>
</div>
<section id="rosalind-string" class="level4">
<h4 class="anchored" data-anchor-id="rosalind-string">rosalind string</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co">#clump_finding("CGGACTCGACAGATGTGAAGAAATGTGAAGACTGAGTGAAGAGAAGAGGAAACACGACACGACATTGCGACATAATGTACGAATGTAATGTGCCTATGGC", 75, 4, 5)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="rosalind-test" class="level4">
<h4 class="anchored" data-anchor-id="rosalind-test">rosalind test</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co">#data &lt;- scan(C:/School/23SPDAY/Biostatistics/clump_string.txt, what = character(), sep = "")</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a><span class="co">#clump_finding(genomeString,511,12,19)"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co">#genome &lt;- rnd_genome(2500)</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co">#clump_finding(genome,500,3,9)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="ecoli-genome" class="level4">
<h4 class="anchored" data-anchor-id="ecoli-genome">ecoli genome</h4>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>ecoli <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"C:/School/23SPDAY/Biostatistics/Group_Project/Genomes/E_coli.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a><span class="co">#clump_finding(ecoli,500,5,9)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
</section>
<section id="using-stringi-and-hash-environments-to-improve-the-efficiency-of-this-code." class="level2">
<h2 class="anchored" data-anchor-id="using-stringi-and-hash-environments-to-improve-the-efficiency-of-this-code.">Using <strong><code>stringi</code></strong> and <strong><code>hash</code></strong> environments to improve the efficiency of this code.</h2>
<ul>
<li><p>To optimize this code, the <code>stringi</code> package is used for string manipulation. Specifically, the <code>stri_sub</code> and <code>stri_count_fixed</code> functions are used to extract all substrings of length <strong><code>k</code></strong> from the initial window of length <strong><code>L</code></strong>, and to count how many times each of these substrings appears in the window.</p>
<ul>
<li><p><code>stringr</code> is a layer above <code>stringi</code> that provides a <u>simpler and more intuitive syntax</u> for many common string operations. However, it is less memory-efficient than <code>stringi</code> when dealing with large datasets. <code>fixed patterns</code> are matched exactly, character by character. This makes <code>stri_count_fixed</code> faster</p></li>
<li><p>In this example, the code is not complex enough to need the improved syntax and relies on only two functions:</p>
<ul>
<li><p><code>stri_count_fixed</code> is a function that counts the number of occurrences of a fixed pattern in a given string.</p></li>
<li><p><code>stri_sub</code> is used to extract all substrings of length <strong><code>k</code></strong> from a given string, which is a crucial step in the calculation of the initial counts of all possible k-mers in the initial window of length <code>L</code>.</p>
<ul>
<li><code>stri_sub</code> is implemented in C and uses <strong>vectorized operations</strong> to optimize performance</li>
</ul></li>
</ul></li>
</ul></li>
<li><p>The <code>clump_finding</code> function uses a hash table to keep track of the counts of all possible k-mers in the initial window of length <code>L</code>. This allows for a constant-time lookup of the count of any given k-mer, instead of having to iterate over the entire window to count its occurrences.</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringi)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a>clump_finding <span class="ot">&lt;-</span> <span class="cf">function</span>(genome, L, t, k) {</span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a>  kmer_counts <span class="ot">&lt;-</span> <span class="fu">new.env</span>(<span class="at">hash =</span> <span class="cn">TRUE</span>, <span class="at">size =</span> <span class="dv">2</span><span class="sc">^</span>k)</span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-6"><a href="#cb79-6" aria-hidden="true" tabindex="-1"></a>  ng <span class="ot">&lt;-</span> <span class="fu">nchar</span>(genome)</span>
<span id="cb79-7"><a href="#cb79-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-8"><a href="#cb79-8" aria-hidden="true" tabindex="-1"></a>  initial_window <span class="ot">&lt;-</span> <span class="fu">stri_sub</span>(genome, <span class="dv">1</span>, L)</span>
<span id="cb79-9"><a href="#cb79-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-10"><a href="#cb79-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"I'm done searching the initial Window!"</span>)</span>
<span id="cb79-11"><a href="#cb79-11" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-12"><a href="#cb79-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(L <span class="sc">-</span> (k <span class="sc">-</span> <span class="dv">1</span>))) {</span>
<span id="cb79-13"><a href="#cb79-13" aria-hidden="true" tabindex="-1"></a>    curr_kmer <span class="ot">&lt;-</span> <span class="fu">stri_sub</span>(initial_window, i, i <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb79-14"><a href="#cb79-14" aria-hidden="true" tabindex="-1"></a>    curr_count <span class="ot">&lt;-</span> <span class="fu">stri_count_fixed</span>(initial_window, curr_kmer)</span>
<span id="cb79-15"><a href="#cb79-15" aria-hidden="true" tabindex="-1"></a>    kmer_counts[[curr_kmer]] <span class="ot">&lt;-</span> curr_count</span>
<span id="cb79-16"><a href="#cb79-16" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb79-17"><a href="#cb79-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-18"><a href="#cb79-18" aria-hidden="true" tabindex="-1"></a>  candidates <span class="ot">&lt;-</span> <span class="fu">names</span>(kmer_counts)[<span class="fu">unlist</span>(<span class="fu">eapply</span>(kmer_counts, <span class="cf">function</span>(x) x)) <span class="sc">&gt;=</span> t]</span>
<span id="cb79-19"><a href="#cb79-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-20"><a href="#cb79-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"I found initial frequent kmers: "</span>, candidates))</span>
<span id="cb79-21"><a href="#cb79-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"Now I'm sliding the window through the rest of the genome!"</span>)</span>
<span id="cb79-22"><a href="#cb79-22" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb79-23"><a href="#cb79-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>(ng <span class="sc">-</span> (L <span class="sc">-</span> <span class="dv">2</span>))) {</span>
<span id="cb79-24"><a href="#cb79-24" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>((j <span class="sc">%%</span> <span class="dv">15000</span>) <span class="sc">==</span> <span class="dv">0</span>){</span>
<span id="cb79-25"><a href="#cb79-25" aria-hidden="true" tabindex="-1"></a>      <span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Hey there! The left edge of my window is currently at nucleotide "</span>, j))</span>
<span id="cb79-26"><a href="#cb79-26" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb79-27"><a href="#cb79-27" aria-hidden="true" tabindex="-1"></a>    rem_pattern <span class="ot">&lt;-</span> <span class="fu">stri_sub</span>(genome, j <span class="sc">-</span> <span class="dv">1</span>, j <span class="sc">-</span> <span class="dv">1</span> <span class="sc">+</span> (k <span class="sc">-</span> <span class="dv">1</span>))</span>
<span id="cb79-28"><a href="#cb79-28" aria-hidden="true" tabindex="-1"></a>    add_pattern <span class="ot">&lt;-</span> <span class="fu">stri_sub</span>(genome, j <span class="sc">+</span> (L <span class="sc">-</span> <span class="dv">2</span>) <span class="sc">-</span> (k <span class="sc">-</span> <span class="dv">1</span>), j <span class="sc">+</span> (L <span class="sc">-</span> <span class="dv">2</span>))</span>
<span id="cb79-29"><a href="#cb79-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-30"><a href="#cb79-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">exists</span>(rem_pattern, <span class="at">envir =</span> kmer_counts)) {</span>
<span id="cb79-31"><a href="#cb79-31" aria-hidden="true" tabindex="-1"></a>      kmer_counts[[rem_pattern]] <span class="ot">&lt;-</span> kmer_counts[[rem_pattern]] <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb79-32"><a href="#cb79-32" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb79-33"><a href="#cb79-33" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb79-34"><a href="#cb79-34" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">exists</span>(add_pattern, <span class="at">envir =</span> kmer_counts)) {</span>
<span id="cb79-35"><a href="#cb79-35" aria-hidden="true" tabindex="-1"></a>      kmer_counts[[add_pattern]] <span class="ot">&lt;-</span> kmer_counts[[add_pattern]] <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb79-36"><a href="#cb79-36" aria-hidden="true" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb79-37"><a href="#cb79-37" aria-hidden="true" tabindex="-1"></a>      kmer_counts[[add_pattern]] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb79-38"><a href="#cb79-38" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb79-39"><a href="#cb79-39" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb79-40"><a href="#cb79-40" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (kmer_counts[[add_pattern]] <span class="sc">&gt;=</span> t) {</span>
<span id="cb79-41"><a href="#cb79-41" aria-hidden="true" tabindex="-1"></a>      candidates <span class="ot">&lt;-</span> <span class="fu">c</span>(candidates, add_pattern)</span>
<span id="cb79-42"><a href="#cb79-42" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb79-43"><a href="#cb79-43" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb79-44"><a href="#cb79-44" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">unique</span>(candidates))</span>
<span id="cb79-45"><a href="#cb79-45" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<hr>
</section>
</section>
<section id="chapter-5-understanding-errors-in-dna-replication" class="level1">
<h1>Chapter 5: Understanding Errors in DNA Replication</h1>
<ul>
<li>As outline in chapter 3 of this notebook, the mechanisms for DNA transcription required to manipulate the building blocks of genes operates in ways that must be considered when</li>
</ul>
<p>The objective is to improve the approach to discovering the replication origin by incorporating biological knowledge. This involves considering the asymmetry of DNA replication and the vulnerability of the lagging strand to mutations, such as deamination, which can lead to a decrease in the frequency of G in a section of the genome. By analyzing the difference in observed frequencies of guanine and cytosine along the genome, a function called SKEW can be used to identify the location to begin searching for the replication origin.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>skew <span class="ot">&lt;-</span> <span class="cf">function</span>(genome) {</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>  skew_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  skew <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genome)){</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>      skew <span class="ot">&lt;-</span> skew <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb80-10"><a href="#cb80-10" aria-hidden="true" tabindex="-1"></a>      skew <span class="ot">&lt;-</span> skew <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb80-11"><a href="#cb80-11" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb80-12"><a href="#cb80-12" aria-hidden="true" tabindex="-1"></a>    skew_list[[i]] <span class="ot">&lt;-</span> skew</span>
<span id="cb80-13"><a href="#cb80-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb80-14"><a href="#cb80-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb80-15"><a href="#cb80-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(skew_list)</span>
<span id="cb80-16"><a href="#cb80-16" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="testing-ggplot-with-skew-using-rnd_genome." class="level3">
<h3 class="anchored" data-anchor-id="testing-ggplot-with-skew-using-rnd_genome.">Testing <code>ggplot</code> with <code>skew</code> using <code>rnd_genome.</code></h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co">#rnd_skew &lt;- skew(rnd_genome(3000))</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co">#ggplot(data = rnd_skew) +</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> position, <span class="at">y =</span> skew))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>mapping: x = ~position, y = ~skew 
geom_line: na.rm = FALSE, orientation = NA
stat_identity: na.rm = FALSE
position_identity </code></pre>
</div>
</div>
</section>
<section id="revising-skew-to-use-a-dataframe-instead-of-a-list" class="level3">
<h3 class="anchored" data-anchor-id="revising-skew-to-use-a-dataframe-instead-of-a-list">Revising <code>skew</code> to use a dataframe instead of a list</h3>
<ul>
<li><code>ggplot2</code> was not a fan of the list, so a <code>df</code> can be generated which will also be beneficial for providing a neater version of the output then the alternative.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>skew <span class="ot">&lt;-</span> <span class="cf">function</span>(genome) {</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize an emptu list to store the value for skew in each iteration of the loop</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>  skew_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>  <span class="co">#initialize a counter to track how skewed the result becomes</span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>  skew <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb83-7"><a href="#cb83-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb83-8"><a href="#cb83-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># if C then -1, if G then +1, update result of skew in skew_list for each iteration</span></span>
<span id="cb83-9"><a href="#cb83-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genome)){</span>
<span id="cb83-10"><a href="#cb83-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb83-11"><a href="#cb83-11" aria-hidden="true" tabindex="-1"></a>      skew <span class="ot">&lt;-</span> skew <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb83-12"><a href="#cb83-12" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb83-13"><a href="#cb83-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb83-14"><a href="#cb83-14" aria-hidden="true" tabindex="-1"></a>      skew <span class="ot">&lt;-</span> skew <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb83-15"><a href="#cb83-15" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb83-16"><a href="#cb83-16" aria-hidden="true" tabindex="-1"></a>    skew_list[[i]] <span class="ot">&lt;-</span> skew</span>
<span id="cb83-17"><a href="#cb83-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb83-18"><a href="#cb83-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb83-19"><a href="#cb83-19" aria-hidden="true" tabindex="-1"></a>  <span class="co">#generate a dataframe instead of a list (for ggplot and aesthetics)</span></span>
<span id="cb83-20"><a href="#cb83-20" aria-hidden="true" tabindex="-1"></a>  skew_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">position =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genome), <span class="at">skew =</span> <span class="fu">unlist</span>(skew_list))</span>
<span id="cb83-21"><a href="#cb83-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(skew_df)</span>
<span id="cb83-22"><a href="#cb83-22" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="generating-min_skew-skew-for-the-e_coli-genome" class="level3">
<h3 class="anchored" data-anchor-id="generating-min_skew-skew-for-the-e_coli-genome">Generating <code>min_skew</code>, <code>skew</code> for the e_coli genome</h3>
<ul>
<li>The <code>min</code> of the skew is where the difference between Cytosine and Guanine is most significant. This is where the most replication is occurring in the genome.</li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co">#gemome "should" be loaded in from the previous chapter</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>e_coli_skew <span class="ot">&lt;-</span> <span class="fu">skew</span>(ecoli)</span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>min_skew <span class="ot">&lt;-</span> <span class="fu">min</span>(e_coli_skew<span class="sc">$</span>skew)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> e_coli_skew) <span class="sc">+</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">mapping =</span> <span class="fu">aes</span>(<span class="at">x =</span> position, <span class="at">y =</span> skew)) <span class="sc">+</span></span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> min_skew, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="dv">10</span><span class="sc">^</span><span class="dv">6</span>, <span class="at">y =</span> min_skew<span class="sc">+</span><span class="dv">5000</span>, <span class="at">label =</span> <span class="fu">paste</span>(<span class="st">"Min ="</span>, min_skew))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="Replication_Trevor_Harrington_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="creating-a-separate-min_skew-function" class="level3">
<h3 class="anchored" data-anchor-id="creating-a-separate-min_skew-function">Creating a separate <code>min_skew</code> function</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>min_skew <span class="ot">&lt;-</span> <span class="cf">function</span>(genome){</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize an emptu list to store the value for skew in each iteration of the loop</span></span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>  skew_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>  <span class="co">#initialize a counter to track how skewed the result becomes</span></span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>  skew <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb86-7"><a href="#cb86-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genome)){</span>
<span id="cb86-8"><a href="#cb86-8" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"C"</span>){</span>
<span id="cb86-9"><a href="#cb86-9" aria-hidden="true" tabindex="-1"></a>      skew <span class="ot">&lt;-</span> skew <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb86-10"><a href="#cb86-10" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb86-11"><a href="#cb86-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(genome, <span class="at">start =</span> i, <span class="at">end =</span> i) <span class="sc">==</span> <span class="st">"G"</span>){</span>
<span id="cb86-12"><a href="#cb86-12" aria-hidden="true" tabindex="-1"></a>      skew <span class="ot">&lt;-</span> skew <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb86-13"><a href="#cb86-13" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb86-14"><a href="#cb86-14" aria-hidden="true" tabindex="-1"></a>    skew_list[[i]] <span class="ot">&lt;-</span> skew</span>
<span id="cb86-15"><a href="#cb86-15" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb86-16"><a href="#cb86-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb86-17"><a href="#cb86-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Convert the list into a dataframe, position = genome length, skew is seperated into individual variables that is the same length as genome position</span></span>
<span id="cb86-18"><a href="#cb86-18" aria-hidden="true" tabindex="-1"></a>  skew_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">position =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nchar</span>(genome), <span class="at">skew =</span> <span class="fu">unlist</span>(skew_list))</span>
<span id="cb86-19"><a href="#cb86-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Find the min skew value found in the dataframe</span></span>
<span id="cb86-20"><a href="#cb86-20" aria-hidden="true" tabindex="-1"></a>  min_skew_value <span class="ot">&lt;-</span> <span class="fu">min</span>(skew_df<span class="sc">$</span>skew)</span>
<span id="cb86-21"><a href="#cb86-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Use the known min value to extract the position of the min skew</span></span>
<span id="cb86-22"><a href="#cb86-22" aria-hidden="true" tabindex="-1"></a>  min_skew_positions <span class="ot">&lt;-</span> <span class="fu">which</span>(skew_df<span class="sc">$</span>skew <span class="sc">==</span> min_skew_value)</span>
<span id="cb86-23"><a href="#cb86-23" aria-hidden="true" tabindex="-1"></a>  <span class="co"># group the min values (if &lt; 1) to paste them with the nice formatting</span></span>
<span id="cb86-24"><a href="#cb86-24" aria-hidden="true" tabindex="-1"></a>  min_skew_positions_str <span class="ot">&lt;-</span> <span class="fu">paste</span>(min_skew_positions, <span class="at">collapse =</span> <span class="st">", "</span>)</span>
<span id="cb86-25"><a href="#cb86-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Nice formatting</span></span>
<span id="cb86-26"><a href="#cb86-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">noquote</span>(<span class="fu">paste</span>(<span class="st">"Min SKEW position(s): "</span>, min_skew_positions_str)))</span>
<span id="cb86-27"><a href="#cb86-27" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min_skew</span>(<span class="st">"CCTATCGGTGGATTAGCATGTCCCTGTACGTTTCGCCGCGAACTAGTTCACACGGCTTGATGGCAAATGGTTTTTCCGGCGACCGTAATCGTCCACCGAG"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] Min SKEW position(s):  53, 97</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>ros_string <span class="ot">&lt;-</span> <span class="fu">scan</span>(<span class="st">"C:/School/23SPDAY/Biostatistics/skew_ros_string.txt"</span>, <span class="at">what =</span> <span class="st">"character"</span>, <span class="at">sep =</span> <span class="cn">NULL</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="fu">min_skew</span>(ros_string)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] Min SKEW position(s):  12860, 12861, 12862</code></pre>
</div>
</div>
<hr>
</section>
</section>
<section id="chapter-6-hamming-distances-for-genome-analysis" class="level1">
<h1>Chapter 6: Hamming Distances for Genome Analysis</h1>
<section id="generating-hamming_distance-function" class="level2">
<h2 class="anchored" data-anchor-id="generating-hamming_distance-function">6.1 Generating <code>hamming_distance</code> function</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>hamming_distance <span class="ot">&lt;-</span> <span class="cf">function</span>(seq1, seq2) {</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nchar</span>(seq1) <span class="sc">!=</span> <span class="fu">nchar</span>(seq2)) {</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"Both sequences must be of equal length"</span>)</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Use an inequality operator to add the numeric value if inequalities found in the indexed value matches in both sequences (as long as they are equal). </span></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="do">## inequality operator can be used in TRUE/FALSE, Logical, or numerical operator.</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>  distance <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">strsplit</span>(seq1, <span class="st">""</span>)[[<span class="dv">1</span>]] <span class="sc">!=</span> <span class="fu">strsplit</span>(seq2, <span class="st">""</span>)[[<span class="dv">1</span>]])</span>
<span id="cb91-9"><a href="#cb91-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(distance)</span>
<span id="cb91-10"><a href="#cb91-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb91-11"><a href="#cb91-11" aria-hidden="true" tabindex="-1"></a><span class="fu">hamming_distance</span>(<span class="st">"GGGCCGTTGGT"</span>,<span class="st">"GGACCGTTGAC"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
</section>
<section id="combining-hamming_distance-and-nt_pattern-to-find-near-matches-with-difference-d" class="level2">
<h2 class="anchored" data-anchor-id="combining-hamming_distance-and-nt_pattern-to-find-near-matches-with-difference-d">6.2 Combining <code>hamming_distance</code> and <code>nt_pattern</code> to find near matches with difference “d”</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>hamming_distance <span class="ot">&lt;-</span> <span class="cf">function</span>(seq1, seq2) {</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (<span class="fu">nchar</span>(seq1) <span class="sc">!=</span> <span class="fu">nchar</span>(seq2)) {</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stop</span>(<span class="st">"Both sequences must be of equal length"</span>)</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Use an inequality operator to add the numeric value if inequalities found in the indexed value matches in both sequences (as long as they are equal). </span></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a><span class="do">## inequality operator can be used in TRUE/FALSE, Logical, or numerical operator.</span></span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>  distance <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">strsplit</span>(seq1, <span class="st">""</span>)[[<span class="dv">1</span>]] <span class="sc">!=</span> <span class="fu">strsplit</span>(seq2, <span class="st">""</span>)[[<span class="dv">1</span>]])</span>
<span id="cb93-9"><a href="#cb93-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(distance)</span>
<span id="cb93-10"><a href="#cb93-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb93-11"><a href="#cb93-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb93-12"><a href="#cb93-12" aria-hidden="true" tabindex="-1"></a>nt_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(string, pattern) {</span>
<span id="cb93-13"><a href="#cb93-13" aria-hidden="true" tabindex="-1"></a>  nt_matches <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb93-14"><a href="#cb93-14" aria-hidden="true" tabindex="-1"></a>  len_string <span class="ot">&lt;-</span> <span class="fu">nchar</span>(string)</span>
<span id="cb93-15"><a href="#cb93-15" aria-hidden="true" tabindex="-1"></a>  len_pattern <span class="ot">&lt;-</span> <span class="fu">str_length</span>(pattern)</span>
<span id="cb93-16"><a href="#cb93-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb93-17"><a href="#cb93-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(len_string <span class="sc">-</span> len_pattern <span class="sc">+</span> <span class="dv">1</span>)){</span>
<span id="cb93-18"><a href="#cb93-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">str_sub</span>(string, i, i <span class="sc">+</span> len_pattern <span class="sc">-</span> <span class="dv">1</span>) <span class="sc">==</span> pattern){</span>
<span id="cb93-19"><a href="#cb93-19" aria-hidden="true" tabindex="-1"></a>      nt_matches <span class="ot">=</span> nt_matches <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb93-20"><a href="#cb93-20" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb93-21"><a href="#cb93-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb93-22"><a href="#cb93-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(nt_matches)</span>
<span id="cb93-23"><a href="#cb93-23" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hamming_distance</span>(<span class="st">'GGGCCGTTGGT'</span>,<span class="st">'GGACCGTTGAC'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a>ros1 <span class="ot">&lt;-</span> <span class="st">"CAGTTAGTCAGGCAAGCGCTAACGATACAAAGTGCCTTGAACAAACACAATAGTACTTGCCGGATCGTTGGTCAGACCATATGAACCTCCGGGTTGAAACCTACAATTGTTGGCGTCCGTGATCCACGGTCTTGAAGTATTTTAGTTCATATGTCTCGTCACGTTTAACTGAGTAATGAAATACATGATGCATTCGCCTCGATGTCTCGGATATTTGGGTGCTCTTAACTGTAATCCTTGCGGTTGTTCTCGCTACCGAAATAAAGCGTAAGCGCAGGGCTGCGCTTAGCTGGGACTAAGTGCCTGTACGTTCGTCTGTGGGTGCGAGAAGCTTCATTTACGAAACTGCGTAGAAGACGGCCTAGCGACACATGAGTGTGCGTGCGAGATAGATGCATTGTAGGGCCGTCTCATGTATCGTAAGTGAGACGATCAGGTTAAGGCGCCCCTCTTTTGGATTGCTGCAAGACTGCTAGCACGATGTCATAACGCTGTGATAGCTATCGGACTAACCGTCGTCCGATTTCTTAACCCTTTGCTGGGAGTGGTGTCGTCCTGCTTTCGCAGGGATTTGAGCGGTCTTGGAGCCGTCCACGGTGTCTACACGTTGGGTTCATTACCCCACAGTTACCAGACGAGGCCACCAAACAGAATGATGACTTCGGAGTTGTAAATCTCGGCCGAGTGGTGCGGTCTCTGGACATTTCATAAAGCGGGATCGACCCATCAGCGCTATGACTAGAGGAGGTCTCTGGTAATTAAGGTATCGCCTCAAATCTAAAAGAATGCGGTTTCTGACACTCGTGTTCGAGAGCTTTATTATTCAAGAATACTTCGTTAACGGCGCGCTGCAAAGTATAGTAGGGGCAGTCGAGTGACGGCTCGTAGAATCGAATTAGACGAGACTATGCTCGGTTTGATACCCCGTACCACGGGTATCTTTTCAGCGATAATTGAGCTTAGTCGTACCGGACCACAACCGAACCCTAAGAGGCGGCGACCGTTTGGTGCCGTCATGCTGTTCTTCCCCCAGGCACGGGCCCTCTACTGTCTCCCC"</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>ros2 <span class="ot">&lt;-</span> <span class="st">"GAGACGTCTACCCTTTCGTTTGTTCTGGTTACGCTACGCTTGCACTGTTCACTGGCCGGGTTGCAGTGCCCGAGCACGCACATCTTAGAGGGTCCGGAAAACCATTAAATTTAGTGCCCATCGATCCGTAGTGTGTTCCATTTATAGTTCCAAACGTATCGGAAGTCTCCACTCAGCTGGGTGGCCCCATGAAAATCAAAATTGATGTGAAACGGCAAGCATGGCAGCAATGGAGTACGGGAGGCTTCGCATGCCCACGTCGGGAGAGAGCGAAGCTCGGTGAGACGGAGGCTTGGGGGGGGTTTTTTATACAGGGCCGCTATAACGAAAAGCTACCTACCAACAATTTACTGTCCAAGCATACCACGGCAACTCATTTCAATCAAGATCTGTAGTCGTCCAGGCACAGCTGAAATAGGTGATCGTGAAGCAGTAAAACTCTCGGGTAAGCTCAGTAGTTACAAGAAATCAGACTGGACTTCTACCGCTGCGATCCAACTAGGGGTTACGGTTGCGAAACTTCGTATGCCCTTCGTACATACGAGGTCATCATATTCCCGAGTCGTTGAGAACTTGATGAGCATCCCCCCACTTGACTAGACATGAGAGTGAACGTCGTGCGGCCGTCACACCGCTAAAGTCGATCCTTCCGAGAACCCACTTTACCGCGGCAAGGAATAGTGCATCTCGACAGGACAGGTGATATTCTGATCCATGAGTCTAACCCGATTTTGCAAAGCCTGGAGCGAAGCGCTACAGGACGCCGGCCCCTTGCCCGTGACGGCCATAGATAGCGGGGCGAGACGTCTATATGAGAGGAAAAGTCAGATAATCCATTATACACTGTGACACAATGAGAACCCTAGAGTAAAAAATCGATGATAAGGGTCTTATTTGGCATGACTACATATGCCCCAGTAGATCCTCGAATGTGCGTTCTATCGGCTCCAAGAACGCTAGGCCAGACAACTCTTCGCTTGACAGGTAAGGATCGGACTAGAACCAAAAGTTGAGGCAGACGAAAAGGGTGAGCGTTAATCTGCCCTGATTGGGGACC"</span></span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a><span class="fu">hamming_distance</span>(ros1,ros2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 819</code></pre>
</div>
</div>
<section id="improving-nt_pattern-to-include-patterns-with-an-acceptable-error-rate" class="level3">
<h3 class="anchored" data-anchor-id="improving-nt_pattern-to-include-patterns-with-an-acceptable-error-rate">Improving <code>nt_pattern</code> to include patterns with an acceptable error rate</h3>
<ul>
<li><p>This code will look through the genome to find strings that match out patterns, but with a given allowed hamming distance between the two patterns.</p></li>
<li><p>The benefit of this tiil will be to improve the</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>hamming_distance <span class="ot">&lt;-</span> <span class="cf">function</span>(seq1, seq2) {</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Use an inequality operator to add the numeric value if inequalities found in the indexed value matches in both sequences (as long as they are equal). </span></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="do">## inequality operator can be used in TRUE/FALSE, Logical, or numerical operator.</span></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>  distance <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">strsplit</span>(seq1, <span class="st">""</span>)[[<span class="dv">1</span>]] <span class="sc">!=</span> <span class="fu">strsplit</span>(seq2, <span class="st">""</span>)[[<span class="dv">1</span>]])</span>
<span id="cb98-5"><a href="#cb98-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(distance)</span>
<span id="cb98-6"><a href="#cb98-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb98-7"><a href="#cb98-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-8"><a href="#cb98-8" aria-hidden="true" tabindex="-1"></a>find_near_pattern <span class="ot">&lt;-</span> <span class="cf">function</span>(pattern, genome, d){</span>
<span id="cb98-9"><a href="#cb98-9" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create learned vector "k" containing the numerical length of the genome</span></span>
<span id="cb98-10"><a href="#cb98-10" aria-hidden="true" tabindex="-1"></a>  k <span class="ot">&lt;-</span> <span class="fu">nchar</span>(pattern)</span>
<span id="cb98-11"><a href="#cb98-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Create empty vector "pattern_locs" to store the locations where the near_matched are found</span></span>
<span id="cb98-12"><a href="#cb98-12" aria-hidden="true" tabindex="-1"></a>  pattern_locs <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb98-13"><a href="#cb98-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print the pattern being looked for</span></span>
<span id="cb98-14"><a href="#cb98-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(pattern)</span>
<span id="cb98-15"><a href="#cb98-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Loop through the genome length and compare hamming distances to see how much </span></span>
<span id="cb98-16"><a href="#cb98-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nchar</span>(genome) <span class="sc">-</span> k <span class="sc">+</span> <span class="dv">1</span>)){</span>
<span id="cb98-17"><a href="#cb98-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># </span></span>
<span id="cb98-18"><a href="#cb98-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span>(<span class="fu">hamming_distance</span>(<span class="fu">str_sub</span>(genome, i, i <span class="sc">+</span> k <span class="sc">-</span> <span class="dv">1</span>), pattern) <span class="sc">&lt;=</span> d){</span>
<span id="cb98-19"><a href="#cb98-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Append pattern locations with the </span></span>
<span id="cb98-20"><a href="#cb98-20" aria-hidden="true" tabindex="-1"></a>      pattern_locs <span class="ot">&lt;-</span> <span class="fu">append</span>(pattern_locs,i<span class="dv">-1</span>)</span>
<span id="cb98-21"><a href="#cb98-21" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb98-22"><a href="#cb98-22" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb98-23"><a href="#cb98-23" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb98-24"><a href="#cb98-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(pattern_locs)</span>
<span id="cb98-25"><a href="#cb98-25" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">find_near_pattern</span>(<span class="st">'ATTCTGGA'</span>,</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="st">'CGCCCGAATCCAGAACGCATTCCCATATTTCGGGACCACTGGCCTCCACGGTACGGACGTCAATCAAATGCCTAGCGGCTTGTGGTTTCTCCTACGCTCC'</span>,<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ATTCTGGA"</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  6  7 26 27 78</code></pre>
</div>
</div>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>