---
title: "Starter Notebook"
author: "You, Scientist"
format: html
execute:
  keep-md: true
---



------------------------------------------------------------------------





## Challenge 1 : Listing four nucleotides


::: {.cell}

```{.r .cell-code}
nucl_names <- c("A", "T", "G", "C")
nucl_names #nucl_names represents the names of the nucleotides. 
```

::: {.cell-output .cell-output-stdout}
```
[1] "A" "T" "G" "C"
```
:::
:::


## Challenge 2 : Creating a random string of 15 nucleotides


::: {.cell}

```{.r .cell-code}
randGenome <- sample(nucl_names, size = 15, replace = TRUE)
paste(randGenome, collapse = "")
```

::: {.cell-output .cell-output-stdout}
```
[1] "TCTAACCTTGAGACC"
```
:::
:::


### Challenge 3: Generate a random genome which is 1500 nucleotides long dataset


::: {.cell}

```{.r .cell-code}
set.seed(215)
genomeLength <- 1500
randGenome <- sample(nucl_names, size = 1500, replace = TRUE)
paste(randGenome, collapse = "")
```

::: {.cell-output .cell-output-stdout}
```
[1] "CGGAACTCCCAACGCCTTGATTCCCGAGTTCTAAGCCGGATCATTGTGGTTTTTGATTGAGAGTCAATCTCAAACGACGTAAGTAGTGTGCGTTGAGCTCTCGCGGATAGGACTATACCGGACGCGAGTTAAGACTCTGAGACGAAAAATAAGCAGGCCTCTCACTGTCGGTCTTAACTACCCCCACTTCCCGTCGTACATCCACGGTTTCTTAATTCCGTGAACCGTGGTACCATGCCTCACGTATGAAAGAGGTATGAGACGCACACTATCTCCTAGTCACTCGATATAGGCAGGTACCGGACGCTAGTGCATGATTGGGGCAGGTAATTTTCCTCGCCGTTTTGTCGGTTGGACGTTAAGGCGCCCCATACTGGCACCCCAAAGACGCGAATCCAAGCAATTCATCCAGTCGATGGGAAGGCGTATACGCACGCGCGATCTCATATTAAGACGTATCCAGGTACTAACAAAAGCCGTTGGCCCGCAACTTCCGATAGCAACTGAGACAACCATCCGTGGGTATACAATTCATCTGGCCTGCTTTTTCCAGTGCATGGGGGGGCGGGGAATTACTAGTGCCTGCACGCCAGCTTCTTGGTACCCCCGGCTATGGTTTCTTACGAAACATACATCAGTGCGTTACCTTGGCTAAACGGTTTCGTGAAGGACGTGGCGTGGCAAGTGCGCGGGTTACATCCGGGTAGACCGAGCCACAAGAAATAGGTAAACCGGGAATCAGCGTGGTATACGCATAGTCTGTATCCTTCGGGGGGTGATCATTGAACCTTCAGCACGCTCGAGCAGTGCATTGGGTTAGTCCCGGGCTTCTCCATTCCGCAACGGGACTGGTTCCACGGGACGTTACGATGATATCGTGGGAGGCCAACAAGCGCTTGTGACAAAGTTCTGCGGCTGAAACTCGCCATGCGTCTCCTCGCTACCCGTTACACCGGCAAAGCCTAAGAGTTATTTCACTCACCGATCTTCAGACTCTTTAAACGCACTTCTTAATAGCATCCTATTCGGACGGCAATTTTATTCTAACTATTATCCAGGCTCTTAGCATCCCCGAGTTGTTCATGTGTCTCTGAAACAAGTCAAGTCAGTCATGGCTTGCGCCAGGTGAAAAACGAACTTTCCCTTAGTGTCATTAAGCGCAGCTGGACGTGGCCGATCCACGTTCGTTTGGCGAGTAGAACCCAAACTGTCGCATTAACTAGTATCATTATAAGGTGAATGGGGTAGTTATGATCGGTGTTATAAGTACGGAAGGAGCGCTGGGTGAAACTAGGGTTGTACCAAGTTGCCCCACTATGGGCAAAGAGGCTATACGCGTATAGAGTTAGACAGCTGGTAGCGAAGGGTATAATGTCCCCCTGGTAGGAAGATTTCCGTATTCCCACTATATTGTGGGCTACTATTGGTTAAAGTAGCTGGTGGCTAATGCCAAAACCACCGTACGCTCAGGCAAAGTAGGATTAGAGAGATGCTGAAACT"
```
:::
:::


### Challenge 4: Generating random genome consisting of 100 nucleotides


::: {.cell}

```{.r .cell-code}
set.seed(215)
genomeLength <- 100
randGenome <- sample(nucl_names, size = 100, replace = TRUE)
table(randGenome)
```

::: {.cell-output .cell-output-stdout}
```
randGenome
 A  C  G  T 
23 23 25 29 
```
:::
:::


### Challenge 5: Writing my own loop


::: {.cell}

```{.r .cell-code}
mySum <- 0

for(i in 1:10){
  mySum <- mySum + i
  print(mySum)
}
```

::: {.cell-output .cell-output-stdout}
```
[1] 1
[1] 3
[1] 6
[1] 10
[1] 15
[1] 21
[1] 28
[1] 36
[1] 45
[1] 55
```
:::
:::


### Challenge 6 :

### Generating a random genome substring consisting of 10 nucleotides using sample(),paste() and loop


::: {.cell}

```{.r .cell-code}
nucleotides <- c("A", "C", "G", "T")
genomeLength <- 10

randGenome <- paste(
  sample(nucleotides, size = genomeLength, replace = TRUE),
                   collapse = "")
print(randGenome)
```

::: {.cell-output .cell-output-stdout}
```
[1] "CTGTGGACAG"
```
:::
:::


### Using for loop to print each nucleotides in my random sample


::: {.cell}

```{.r .cell-code}
for(i in 1:nchar(randGenome)){
  
    print(str_sub(randGenome, start = i, end = i))
  }
```

::: {.cell-output .cell-output-stdout}
```
[1] "C"
[1] "T"
[1] "G"
[1] "T"
[1] "G"
[1] "G"
[1] "A"
[1] "C"
[1] "A"
[1] "G"
```
:::
:::


### Challenge 7 : Counting the number of occurrences of Adenine (A) in randGenome.


::: {.cell}

```{.r .cell-code}
sum_A <- 0

for(i in 1:nchar(randGenome)){
  if(str_sub(randGenome, start = i, end = i) == "A")
    sum_A <- sum_A + 1
    {
  }
}
print(sum_A)
```

::: {.cell-output .cell-output-stdout}
```
[1] 2
```
:::
:::


### Challenge 8 : Counting the frequencies of each of the four individual nucleotides.


::: {.cell}

```{.r .cell-code}
sum_A <- 0
sum_C <- 0
sum_G <- 0
sum_T <- 0

for(i in 1:nchar(randGenome)) {
if(str_sub(randGenome, start = i, end = i) == "A"){ 
    sum_A <- sum_A + 1
}
  if(str_sub(randGenome, start = i, end = i) == "C"){
    sum_C <- sum_C + 1
  }
  if(str_sub(randGenome, start = i, end = i) == "G"){
    sum_G <- sum_G + 1
  }
  if(str_sub(randGenome, start = i, end = i) == "T"){ 
    sum_T <- sum_T + 1
  }
}

print(c(sum_A,sum_C,sum_G,sum_T))
```

::: {.cell-output .cell-output-stdout}
```
[1] 2 2 4 2
```
:::
:::


### Challenge 9 : Counting the frequencies of each nucleotide in Vibrio Cholerae chromosome


::: {.cell}

```{.r .cell-code}
vib_c <- scan("~/Desktop/vib_c.txt", what = "character", sep = NULL)
sum_A <- 0
sum_C <- 0
sum_G <- 0
sum_T <- 0

for(i in 1:nchar(vib_c)) {
if(str_sub(vib_c, start = i, end = i) == "A"){ 
    sum_A <- sum_A + 1
}
  if(str_sub(vib_c, start = i, end = i) == "C"){
    sum_C <- sum_C + 1
  }
  if(str_sub(vib_c, start = i, end = i) == "G"){
    sum_G <- sum_G + 1
  }
  if(str_sub(vib_c, start = i, end = i) == "T"){ 
    sum_T <- sum_T + 1
  }
}

print(c(sum_A,sum_C,sum_G,sum_T))
```

::: {.cell-output .cell-output-stdout}
```
[1] 293942 263573 256024 294711
```
:::
:::


### Challenge 10 : Counting the frequencies of each nucleotide in Vibrio Cholerae chromosome


::: {.cell}

```{.r .cell-code}
vib_c <- "GGCACCTGATGCAGTTCAAACACGTTGAAGATAGTACTTCCAATCTGCCCTGGAATGCCCCTCTGGTGGTACATATCATGTAGGGCGGAACCCGCTTTAACGCCCCTAGGACAATGAAACAGACGAACCGCTTACAAAAGGGGAATAGATGTGACCCTCTCGTAACTGCTCCTCACATTAGCTGAGTGGGAGGCTGTGTTTGTCGGGCTAAGCTTCGCTTGCCAAAATGAAGGATTCTTTAGGGATCGCACTGAACAAGGGTGAAGATCAGATGCGATCAGCATTTCATTGCCTTCTGAATTATAGATGACCCTGCTTGTACTGCTCACGTGGACGCCATTCAAAGCCTTGCGACCAACTCGGTCCAATGTCCCAACCTTTTTCTAACGATCATCATGACTCGGGCAGCGCAAAGCGAGATCGCTAGTTCATTAGTTCCGACCGTCGTAAGACTCACAAGGACGAGGCCAACTCACTCATCTGTACTAACTTTCCCAGAGAGCCTTGATCTGGTGCGACGAGTAGTATAGCGCCGGCTCTGACGTGGTGCACGCGCCACCTATTGGTCGCTCTGATCCACGTACTAGATCATTTAAAAAGACCTGTGACGCTAAAGATATAAGCGCTATATAGTACCTTTCGTGACCCTAGGTTAAGCGACTTGTGACTGATCCTGAAGTACAAAACCTTTGTGCATGGAGTCAGTCATTCCTTTGCTCTGACGTAGTTCTCGTTGCTTAGTAACATCTTGCTCTGATTATCGCGTACGGATACGCTCTCTCGCGCTCAGACGGATGTGAGGTTCGGTGAAGGGTGAGGTCGACTGTGAATTATGCGTGCCGGGGTCGTTTAAGCGTTCGCATCCGGAGCAATGGCGGATAGCTCAAGTGGTTAGGTGGTACACTGAACGCGACGGTCTTTC
"
sum_A <- 0
sum_C <- 0
sum_G <- 0
sum_T <- 0

for(i in 1:nchar(vib_c)) {
if(str_sub(vib_c, start = i, end = i) == "A"){ 
    sum_A <- sum_A + 1
}
  if(str_sub(vib_c, start = i, end = i) == "C"){
    sum_C <- sum_C + 1
  }
  if(str_sub(vib_c, start = i, end = i) == "G"){
    sum_G <- sum_G + 1
  }
  if(str_sub(vib_c, start = i, end = i) == "T"){ 
    sum_T <- sum_T + 1
  }
}

print(c(sum_A,sum_C,sum_G,sum_T))
```

::: {.cell-output .cell-output-stdout}
```
[1] 222 227 233 240
```
:::
:::


I copied the whole dataset from Challenge 9 and pasted it into challenge 10 by removing the chlolera dataset. I copied and pasted the ACGT sequence generated in Rosalind. I checked the results in the website and they came to be correct. The results were A: 222, C: 227, G: 233, T: 240

# **Finding the Replication Origin, Part I**

### Challenge 0: The search for the Replication Origin


::: {.cell}

```{.r .cell-code}
nucleotide_frequency <- function(genomeString, nucleotide = "A"){
  count <- 0
  for(i in 1:nchar(genomeString)){
    if(str_sub(genomeString, start = i, end = i) == nucleotide){
      count <- count + 1
    }
  }
  return(count)
}

nucleotide_frequency("ACTTGCGGGTATCGAG", "G")
```

::: {.cell-output .cell-output-stdout}
```
[1] 6
```
:::
:::


## Challenge 1: Using the sample() function in a randomized genome


::: {.cell}

```{.r .cell-code}
nt_names<- c("A","C","G","T")
nt_sample <- sample(nt_names, size = 2000, replace = TRUE)
nt_sample <- paste(nt_sample, collapse = "")

nucleotide_frequency(nt_sample, "C")
```

::: {.cell-output .cell-output-stdout}
```
[1] 487
```
:::
:::


## Challenge 2: Building a function to create a randomized genome


::: {.cell}

```{.r .cell-code}
rand_genome <- function(k){
  nt_names <- c("A", "C", "G", "T")
rand_genome <- sample(nt_names, size=k, replace= TRUE)
  rand_genome<- paste(rand_genome, collapse="")

  return(rand_genome)
  }
rand_genome(k=15)
```

::: {.cell-output .cell-output-stdout}
```
[1] "AACAGTAAATGACAT"
```
:::
:::


## Challenge 3: Building a function called to generate all of the substrings of 3 nucleotides


::: {.cell}

```{.r .cell-code}
generate_3_mers <- function(myString) {
  list_3_mers <- c()

  for(i in 1:(nchar(myString) - 2)){
  list_3_mers <- list_3_mers %>%
  append(str_sub(myString, start = i, end = i + 2))
    }
  return(list_3_mers)
}

myGenome <- rand_genome(10)
myGenome
```

::: {.cell-output .cell-output-stdout}
```
[1] "TAGAAGACAA"
```
:::

```{.r .cell-code}
generate_3_mers(myGenome)
```

::: {.cell-output .cell-output-stdout}
```
[1] "TAG" "AGA" "GAA" "AAG" "AGA" "GAC" "ACA" "CAA"
```
:::
:::


## Challenge 4: Generate k-mers of any length we want


::: {.cell}

```{.r .cell-code}
generate_k_mers <- function(myString, k = 4) {
  list_3_mers <- c()

  for(i in 1:(nchar(myString) - (k-1))){
  list_3_mers <- list_3_mers %>%
  append(str_sub(myString, start = i, end = i + k-1))
    }
  return(list_3_mers)
}

myGenome <- rand_genome(10)
myGenome
```

::: {.cell-output .cell-output-stdout}
```
[1] "GATGCAACGA"
```
:::

```{.r .cell-code}
generate_k_mers(myGenome,6)
```

::: {.cell-output .cell-output-stdout}
```
[1] "GATGCA" "ATGCAA" "TGCAAC" "GCAACG" "CAACGA"
```
:::
:::


## Challenge 5: Write a new function that will count occurrences of a particular pattern within a larger genomestring.


::: {.cell}

```{.r .cell-code}
count_pattern <- function(myString, pattern){
  count <- 0
  for(i in 1:nchar(myString)){
    if(str_sub(myString, start = i, end = i+1) == pattern){
      count = count + 1
    }
  }
  return(count)
}
  
count_pattern(myGenome, "AT")
```

::: {.cell-output .cell-output-stdout}
```
[1] 1
```
:::
:::

::: {.cell}

```{.r .cell-code}
rosalind_string <- "TTTATGAACTATGAACTTATGAACGTATTATGAACTATGAACTACACATATGAACTATGAACTTATGAACTTATGAACTATGAACTATGAACCTATGAACTATGAACCCTTATATGAACTACATATGAACTATGAACTCTATGAACGCACTGGTATGAACCTATGAACAGTATGAACATGCGATATGAACAGTCACTATGAACTAGGCTGTATGAACTATGAACCTATGAACTATGAACAACGTAAGTATATGAACAACGGTATGAACGTATGAACTATGAACCAACGTATGAACATATGAACTAGTATGAACGGGATATGAACACGGGTGTATGAACCTCTCTATGAACTATGAACTATGAACACTATGAACAGTATGAACTTATGAACTATGAACGATCAGTCTATGAACTATGAACCTTTATGAACTATGAACTGAATATGAACACACTCCTATGAACCTTATGAACCTAGGTTATGAACGGAAACCGGTTATGAACACTTATGAACTATGAACTATGAACCCTTCTTATGAACGTTATGAACATGAGGTTATGAACTTTATGAACGTTATGAACCTTTATGAACTCCTATGAACTTATGAACTATGAACGTCATATGAACGAGGACGATATGAACGCTATGAACTATGAACTCGTATGAACCTTATGAACGTCGTGCTATGAACGTATGAACAGATGTATGAACTATGAACCGTTCGCTATGAACTCTGTTATATGAACTATGAACGGGTATGAACGTATGAACAGATATGAACTATGAACGAATATGAACGGTTGTATGAACCTATGAACTATGAACTATGAACACTTATGAACGTATGAACATTTATGAACCGGGAATATGAACCAGAACTTATGAACATATGAACTATGAACATCAGTATGAACATATGAACATCATAAGTTTATGAACTATGAACGTTTATGAACGCTATGAACTATCCGATTTATGAACGTTTATATGAACTATGAAC"
count_pattern <- function(myString, pattern){
  count <- 0
  k <- nchar(pattern)
  for(i in 1:nchar(myString)){
    if(str_sub(myString, start = i, end = i+k-1) == pattern){
      count = count + 1
    }
  }
  return(count)
}
count_pattern(rosalind_string, "TATGAACTA")
```

::: {.cell-output .cell-output-stdout}
```
[1] 32
```
:::
:::


# **Finding the Replication Origin, Part II**

Challenge 1: Finding frequent


::: {.cell}

```{.r .cell-code}
find_frequent_kmers <- function(genome,k){
  #Get the K mers
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  #Count occurences
  kmer_counts <- rep(0,length(kmers))
  for(i in 1:length(kmers)){
    kmer_counts[i] <- count_pattern(genome,kmers[i])
  }
  max_freq <-  max(kmer_counts)
  #Get most frequent kmers
  freq_kmers <- kmers[kmer_counts==max_freq]
  #the two equal signs ask whether the left hand side is equal to the right hand side
  return(freq_kmers)
  
}


find_frequent_kmers("ATCCATTAT",2)
```

::: {.cell-output .cell-output-stdout}
```
[1] "AT"
```
:::
:::

::: {.cell}

```{.r .cell-code}
rosalind_string <- "ACGTTGCATGTCGCATGATGCATGAGAGCT"
find_frequent_kmers <- function(genome,k){
  #Get the K mers
  kmers <- generate_k_mers(genome,k)
  kmers <- unique(kmers)
  #Count occurences
  kmer_counts <- rep(0,length(kmers))
  for(i in 1:length(kmers)){
    kmer_counts[i] <- count_pattern(genome,kmers[i])
  }
  #print(kmer_counts)
  max_freq <-  max(kmer_counts)
  #print(max_freq)
  freq_kmers <- kmers[kmer_counts==max_freq]
  return(freq_kmers)
  
}
find_frequent_kmers(rosalind_string,4)
```

::: {.cell-output .cell-output-stdout}
```
[1] "GCAT" "CATG"
```
:::
:::


## Challenge 2: Using Reverse complement function on a genome string


::: {.cell}

```{.r .cell-code}
reverse_complement<- function(substring){
  #reverse the substring
  substring_list<- str_split(substring,"")[[1]]
  #separates the character strings into individual variables
  rev_substring<- rev(substring_list)
  #Reverses the base pairs 
rev_complement <-c()
  for(i in 1:length(rev_substring)){
    if(rev_substring[i]=="A"){
     rev_complement <- append(rev_complement, "T")
    }
    if(rev_substring[i]=="T"){
      rev_complement <- append(rev_complement, "A")
    } 
  if(rev_substring[i]=="C"){
     rev_complement <- append(rev_complement, "G")
  }
  if(rev_substring[i]=="G"){
      rev_complement <- append(rev_complement, "C")
    }
  }
  
  rev_complement <- paste(rev_complement,collapse="")
  return(rev_complement)
}


reverse_complement("TAGGTC")
```

::: {.cell-output .cell-output-stdout}
```
[1] "GACCTA"
```
:::
:::


using the rev_substring, we can flip the nucleotide from 5' to 3' or 3' to 5'

### Rosalind String


::: {.cell}

```{.r .cell-code}
reverse_complement<- function(substring){
  #reverse the substring
  substring_list<- str_split(substring,"")[[1]]
  #separates the character strings into individual variables
  rev_substring<- rev(substring_list)
  #Reverses the base pairs 
rev_complement <-c()
  for(i in 1:length(rev_substring)){
    if(rev_substring[i]=="A"){
     rev_complement <- append(rev_complement, "T")
    }
    if(rev_substring[i]=="T"){
      rev_complement <- append(rev_complement, "A")
    } 
  if(rev_substring[i]=="C"){
     rev_complement <- append(rev_complement, "G")
  }
  if(rev_substring[i]=="G"){
      rev_complement <- append(rev_complement, "C")
    }
  }
  
  rev_complement <- paste(rev_complement,collapse="")
  return(rev_complement)

}
genome<-scan("/Users/darvinojha/Desktop/rosalind.txt",what = "character", sep = NULL)

reverse_complement(genome)
```

::: {.cell-output .cell-output-stdout}
```
[1] "CCGATCTATTAGGAACCGTAGCGTCTCTGCGTTTAAGTCCAGAAAATCGCACCTCCTTCTAGTTCGCGATTGGCAGTACCGTGAGGTGCACAAGATCCGTTCGCGTCCGAAGCCCGACGGTGTCTCGTGTGTGAGACCGCGACCAGACTGATCCTGCCAACAATGCCCCTGATACTCACCGATGACTATTTTAAGTATTCGTTCTATAGAGTCGTGTTCCCATCAGGTCAGCCCCTTCGCCCATCCTGCACGCATTTCCACCGAATTAACTGACCTTATTTAGAGGGTAATGCCCATCTTAATAACCAATGGGACTAGTGACATGTTTGGTCCGTGTATATCAATAGAACCGGCAGCTAAGGCCTCGGTTGTTGTCCGCGCCTTGTGGAGATGTGAGTACAGTGCCCTAGCGGTTCTGCTGCATGTAAAGGCAGTGCGGTCTCCTGTGTCCCGTTGGACTTTGACAAGACTACGGTGAGTGTAAAAGTGGCACATCTCACAAAATATAGAGCGAAGCTAGGCTCACCAAGTCCGGCCTACCCACATGGTGATTTAGTAGGAAGTTGTGTGTGATCTGATGGAGCGCCAACGATATCTACAACTATATTTCGACTAAGGGTACCTAGCAATTCGACCCTTATCATCCACGAGTTAACGTTTGATTACAATCCTCCAACACCCTACCCCTTGGAGTCTTTGTGTATGTTATTATGCGCTGGAAAAGGACGCGCCATCTCACGAGTCGCCCAATTTATGAGCCGGGCCCCTCCGAACGCGTGACGGGTGCAGAGCTGTTGAAGTTGTCGGGAGCATTAGCATATGCCCGTCACATAATTGGGTCCACCCTGATACACCCGCTTGTCAGTCCCTTTTGATAGCCACACATTTGTTGGGAGGTCTTGGCGGTTCCCGATTTACTCAGATTTCTCATCGTGAATGTTACGGTGTAGAAAATCAATTTCTGACCCGGACCTATGGCACGAACCGGCCGCGAGTACCCCGTAGATTAGACCCTGGGTGCTGAGTGGACTCCGCAGGGTCGATTACAGTCCTAGTGATCTGAGTGTGTACTGATCCAGCTTGGATCGCTCGAAGTTAATGTTATTATGGCTGTTGGAGGTTGTAGACTTAGGGTCTAAAACCATCGACGTTTATAACCTACTTAGAGATGAGTGTTTTTTCTCTATTATACCCAAATATATTCAGTTGTTGTGGTGTTCCTCACCAACTAACGATACCCGAATCTGAAATCTAACCCCAGGAAGCCTCTAGGAGATAACAAGGTCAACGTTACGTGAGTGGACCCGCGCTGCTCGTTCATAACCTACGTATCTCTTGGCTGACGGCGGGTGAGAATACCAATAGCTCAACGAACGATCGCAAGCATAGCTAAACATACCACGGCGATCTCTAATGCCCAGTCAACTCAGCGACTCATCGGAGTTTGCCTTTTGGACCGGGGTTAGCTTACGCGTTTTCCCAGGAGGTCGGAGGGCAACCTGGTAGGTTTCACCCAAGGAGAGACGGTATCGGTATGGCGACACGTTTGAGTCATGACGGTCCACCCAATCGTAAGTAGATTTCGTCATGTTTCTGTCGTTTAACCTCGGAGCGTGTCGCTATCTATTCATTTGTTAGCCTTCCGGCGCAATATTCACTCACAAGTAGGCGCGATGGAAAGGAGGTCGACAGGATAACAGATAGTCCATGTGGAACCAGACTAGTTATGTCATCTTACTGCTAACTTGACCGCCATTGAGGGTAGAATCGGTATTACTGTAATCCAACCACGGCAGCCCTGTTCACTCATAGGGTACTGCGCAGCACGTGGCCATGAGCGATGCCCCCAGGAGAGCTGATTTTCATCCAACTGCAATCTAATAGACTAGACCTAGTGCCAAGACTCATAATGGCTCTTAGCAGGGCTCAGCAAGAAAAGTGTCATACTGATCCTTCTCTAGATAAGTCTGATAGCATCTGTATTGCGGAAGCGCAACGAACGTTGACACACGGGAGGCCAAGCTATAGGAGTTCATGGACGCCCCTTCGTATCGTAGGATAGGTAACGGTCATCTTCAATACCCCCACTGTGCAAGCTAACATGCGAATGAGACCGTGATTTGCAACTTAGACCACTTCCGGCCAGTCGACAAGTAGAGATGAGACGAGGTCTCTCCGATCGAGTCTGTGCGGACTGGTTCATACGAGCTTACTCAGTGTAAGCATCGCGCTACAAGGTAGCGGGCCTGGCCACGCATTAGCGAAACAATGCACCTGAAAGACGCCTTCGCCTAGGAGACTTCGGTGTGGTGCGACGGCCTCCGATAAAACATTAGATTGCCGCCGTGTTACTTGAACTAGTCCCGTAAGTCGCGCTTAAAAAATAGTGGGCGTACTTTTTTTAGGAGCAATGGTCGTCCGAGAACCCTATTCGTCACTGGGGGTGGGGCCCACCTGAGACTAACCTGGCGCTGGAACCGCTGACCCAGTCCTAAATTTTCCTACCGCATAACCATAGCACCCCATCGCGATGCTTGCATGATCTCGTGGGATCGTGACTCGGGCTACAGCCATGGTCTCTTATTCTCTTGATACTACCTGCGAACTTTGATGAACCGTTTTTAACCGCGCAGTAACGGCCATGTATAGTTTGGTATTAGACAAGATGAATGGGAGCCAGGACGTTAACCCGGTCGACATTATCAGCGTTGACGCAATATGTGGGGATGAGTCGGCCCATGATCACGAAACCACACGAACGTTCCCTGTGTTGGCTCCGATATGGCTTGTGAGGTGGGCATAAAGTCCTGACTCTGCACGGACTGAATCCATTCTAAACCACACTACGGAATTCCGCTTTCTATTTTAGATGTTTATATAAGCCCCGAGTGGGTTGAGAGTGTAGTGGCTGGGCAGCCTGTCAGCATTTCGTTGGCGTACCAGTATACAGCAAGGTGCACACCGGGAGTAATCGCTAGAGAATATTCCATTGCATGCACGTATAGTCTTCGATGGCCGCAAAAAATACAACCTCTAGTTCGATGTTCTACACGCTGAGACTCATCGGTATCGCTGAGAAAGACATCGAGTTTTACTATGCCCTTCTTAACGGGATTATCTATGAGTTTGCATACATACATATACCCATGGCTTGTCGTATAGTCTGCGTTCACTAGCAGCGCCTCTGCCTCGTAACGTGTAGTTCTCGGGGCTACAGTCGGTCATCCTCTAGGGCCCCATCATGGCTTGCGTCGTTTATCAGAGGCGTCGGGCAGATTTGGCCCGCTATCTCATGCCTAGATCCGTCTCTGTAATACAAGCCTCTTGACTAGTGTCCGAGGTTATGGCCTAGTTACCTTGGCCGTGAGGGTGGTTCTACGCAGCCAAGCATGAGGCTCACGCATTAGGACCGAAGGGTATTAGCTTTTAAAATAATGCAACTCCAAGTGCCCTGTACGTTGGGCACTTGGCGCATGAGAAAAGACGTCAAGGGAGATGTAACAGAGGTCGGGTTCCAGACTAAGAGGCAGTGCCGCTTCGACTAGCGTCCTGTTTGTATTACGACTCCCACGACCTGGTAGATAACAGGGACGGAGACTGCTGAAATGTGCCGCGGAAGGTCCTGTTTGGTTTAAGGGCGGGCGCCTCGGCGACGGTCGCAGACGAGAAACTTCTGTTAAAGATCTTGACTACATGCTAGTTCCCCCTCCGGGCGGAAATCGTTGGTGTTAACTTGTATTCTTACAAGGCACCTATCAACGACAGGAGTTTCTTTTTTCCTACGCTGAGAGTGAGGTATCCTCAGTTGACTTAAGTGAGGCCGTTAGTTGGCACAGTGATTTGTGCCGACAACCGACCTGACTTGCCTACCCGCACCTTTCACGCATGGGCTTTTCCGGCCCAGAGGCGAAATCATTGGAGTATATCATAATCACCAGGGCCCGAGTTCACTACAAGGACCCCTATGCCCATACAGAGAAAACGTAATTGCGCGCTTAGCCTCAATGCAGAGTGCTTTTATTGCGGACCGTGTGCAGCTAGCCAACTCTTGTACATGAGTTAGCGAATCACTGAATAATGGGACTTTCGTGATTCCGAGTTCTCAACATAAGGGTGGGATTCGATTGTTCACACTCGTGAGTAAGGCTTTCCTAGTGGTGCATGAGATACCAGGTAGACCAGCCTAGAAAATGATGGCTGTTTCCTCGTTTCCCCATAAGGCTGGAACATAGACTAAAAGCAGCTATCGTGAGGCTCCACGTCTGTACCAGGAGCGCCCCGTTGTATTCCCTCAAGTGGGACGAGGCCCTACTGCAGGATGACGGAATGCTGTTGGGTCGTTGGTATCTAGTATTCCGTGCGGCTCTTCCTAGAGATCTGGGCACCATCCGAAGAAGACCTACAAACGGCGCCATTGCACCATAGTACTAATGTTGGTCCAGTAAGTGCGACTGCACGCCTTACCTTCCTCTCCCTTAACATACATAGCAATATTTCGGGATCTTCCAAGCTAATCTATGTAGGGCTGTGATTTGGTGCATTGCGCTTCTCGTACATCGCGGACATCGAAGGAATACCATGATATGACATCCGGCTCCACTCTGGTGAGGGGAGCTTGGCTGAGCTAGCTCCAACGGTCAGCCGGGGCGATCAGCCGACGAAGTCCAGATACGATCTATACAGCCAAACTAATGCGGTATAACAAGCCACAGGTTACAGCCCCCTGTTGTGGAAGACGGTTCCAAATAGGTGGACAAGTGAGCGAATAGTCCGCCCTTGGTACAGATAACGAGCGCGGTTTTACGAGAAAAGGCGACAGCCGCATCACGTACCGGTCGGGCGAGAATGAAAGATACGTTTCCCGAGGGAGTGTCGCGGATGTAAAAGTAAACATTCTCGACGGTTCCGGGCGGGCGTCAAAGAGAACCTACTTAGCGGGCCAGTTATGCTCCTTTGTCGGCTTTCACCGGTCCGGCCGGGCGCCAACCACAAGGTATACTGTGAGCACAGTTGGTTAGAAAATAAACGCGATCAGTTAAACTTGATTCCATACTATGACCGTCCCTTGCATCGTCAGAGAGTATGGTTCGACGCGTACACGGACCCGCAGCGGGGACCGCTATGACTGGGTCGAGATTTGCACAAATACACAGACTTCCCATGGAGGACAATGGTCCGGATACTATGTAAGACAATATTACACCATCTGGACGCCGTCCCGTTCGCCCGCGGATTCTCAATTTCTAACTATAGGTCCGTGAATGGCCGAACTGCCGCGAAAGCAAGGCATACAGGTGGGGATGGCTATTGTCAGGAGAGAGAGGTGGCCGCGCAGTAACGTTGTCTGATCCTACTGCCTGACTTGTACGACCATTTATAAATAATCGATCAAGACCGTACCGGGAAACCCTCACTTCCTGCAGTTGTTGCTGCTCTCTGGATGGGTGTCGAAGGCCCCATGGATACGACCCCTCCTATTGGCCAGAGAGTATAGCCGGTTACAGTCTCCACTCAATACCACGTACGGACAGCATGTGGCGCCTACGGGGCTCTTGTGTATAGACTACATATTTATATCTAGTTGAAGGCAGCACTTTACGAACAGTACGAGGAAAGTTATTCTGAGGGACGGAATTCAAATTTTTTCGACACATCCCTCCCGCGAGAACGCTTGAACCCGACAAGCATTATCTCCCAACGTGACGGATCGGCAAGGGGGCGTAATGGACAGGACCCGATTCCCTCCCGCTCCAAACGCTAGCACGATTAGTTAAGAGGTAGAGTCGTTCTTCATACGTTCGGGAATTAAGCGTCACGATGGTGCGTATAACAGCTTCGCACTGGATACGCTTTACCTTTAATTATCGGATCTGTTCGTATGCCGTAATGAGTAATACACCTGAGTCTATCAGCCGGCTACGGCCTCGGCGCCCTAGGGACAAAATTCATTCTACACCGGGGGCCTACGCATGTGAGGGATTAGTGCTTGAGCGGGCCCTCCCTCCGGCCTAAGTTTTCGCGCCACCGGCGCCCGGCAAGTATAGGGCTACCGATCAATGTTGCGATATCTCCCGCAAGTACGCGGAAGACCGCAATTCATATTGAAGAAGCCCAATCGTAGCCGGCTTTGAGGGACCACAAACTGCCTTTTTTGAAAGGTAGATTCTCAGCTAGCCGTCAACGATTTTAGAATAGATGGTTATGTCGCCGAATCAAGACGTGCATACACTACTTCGAGCCGATCTCTTGAGTACGGTTATTGCCTGCGTAGTGCTCAATCGCTTCTCCACTGCACGCTAGAACACTGTGGGTGTTTATGTGGTCCGAACTGGAGGGCATAGTGCGGTGTGGGGATCGGACAAACGTTCGCCGTATATCTATGGAGTTGCCACTGACGACGCATCGTGAGACCGCGGAGACGTGCGCTGATCGATCGGTGACTCAGGAACCACATCCGTGGAGTTTTTCCGCGTCTAACGTTTGCTGGCCTTTTATCTTTGTCCAGCCTACAGGTGCCAGAGGTCTCATCCGATAATGTTCCGTAGAATCCCATAGTGACTTAGTGAACGGGCGACGTTGCACTCAGGGAACAACACTCATTCGTCTGCGTACGCTCGTTTTGCAGCGTACCCTGGTACCCAACAGATAAGAGTACCGTTCTGCTGGGCATGATGAATTGGACCGGTCTGACACTACTCAGGCCAAGTGTCAAACTCATCCAAAGTTGTACGTCTTATGTATTGGATCGAGCACTCCGCACGAGTAGGCCCTGAACAACTCATGCTATCCTACAAATTTTAGTGAGTCCTCACGCTGTACGCGCAACTAACAATCCAGGGCGCCGCGGAACATGGCTCCCACGACACGTCCACTGAAATCCATGGGGTACACTACCATCCCTTATCGTTCCGTTCTACTGCCCTGGGCGGCAATCTTGAGCTGAGCCTTACCAAGCCGCCAGCCAATGGCTACATATGTGGGCAAGGCTGAGGAGTTTAATTGCTTTGGATGGAACCCTTACTCAGATACCATAAGTCCCATAGGGAGACGGCTGTCTACTTGCTTCGATCGTATGTGGACGGTCCTTCGCCGCTGTCTGTCCTTCTGAGGATGACTGCGTTTTAGCCACGGCCATACGAAGCTACGAACGCCAGAACTACGGCACCACTTATGAAATGGGCCCTTACTGACGAAGGGTGTATCGAGGTGTTAGCACCCACCTACGACGGATCACGTTGCCATGAAGCCAGGCAAGGAAGGCGAACCGTCGTCTAAAGGTAACCAGGTATGTAACGGTCCAAGTCAATACGTCGAGACTAGGCAGCACCCAAGCCGCTCAACTGTACAACATCTGTTAGGGAAATCCCGTCGAAACAGTGTCGTATATGTTAAATACGTTCGGATGAGAAAGTTGGAAATTCTTTCGTCTGCCAGGCCCAAAATAGAGAGTTCCCGTGCCTTAAAAGCGCACCACGTTCTTGCGACCTTCATCGGGCGTAGCAAGTAACTTTTCCGTTTCAAGGGTTATACCCTCACTCGGCGAAACATGATGCGTTCATGGCACGGCATAGGGGTATTACATTCGATAGGGCCCAGTGAGATCATCTCCAACCCGATCGAGCATGACTGTGTCTTGCCGTACATTATTTTACTTGATAATTGGTGGGGATGTTACCGGCACACCCACAGCGAAGCATGCGGGCCGGGGTACTCGCAATTCTTGATCAACATATACAGGTTTACTTGAGATTGTGGTGGGTGGGTCCTAAATATGGTGCTCGACGCGGCATGTGGATGTATATGTCCTGATTTACTAGCTACGCGAGGCGCCATGTTTCGATTCATCGCGGGCTGGGGGGAATTAACGATACGTCAATTATTCCGTTTTGACATAGCTAATCGCCTTACCAGCACTCTTACACAAGACGCTTATCATTTGTTTCTCACCGACCCTCAATTTCCTTGGAACCAGATGAATAACACTATCGTGCAGCCTACTGTGCACTAAAGTCCGCCTGGATCCCGCGGGAATGTCTCAATGCATAGTTTAGCCTTTCAGGCGCGGTACCAGATTGCAGCAAGAGACTGGAAGGGCGTCTCGAAGTGAGTGAGTTGTACACTACGCGGCGACCTACTTCTGTTACCCCTCCTGCGTACAGACCGAACTTAACCACTAATAGGGGCAGATCGGGACCTCGGCCCGGTCAGGTGAGCCGCAACTGGCTCGTCCGATCAGGGATATAAATAATTACGAGATTCGAGATTGCCCGGCCTTGCAATCGACTAACCACCTGAACGAGTTATCATAAACACGCCTAAGTTATTATGATTTACGAAGCACGCGCTCTGTCTACAAGCAGCGTCAAAGAGCCGATCTTGCGGGATTAGACTCTGTAGAGAGGTTCAGAATTCCGGGGAGTTCCTTCCCTATTTGGAAGTGAGGACATAGGGAGAGTCCCATCAGGATCCGCGTGCAGTACACCGACAAGGAACTCTTAATCCTTACGCAAGTTTCGCATCAGATGTTATCTTCTTGGATTACTAATAGGCGCCTGTATGTGACGCGCACAGCGAAGATTGCGAGGTTCCTAGAGAAGTCGTTTTCTAAATAATATCTACAACGGTCCTACAGTATTATGTGTATCTGGAAAAAACGGTCAATGCCTTTATGATTAGATGACCCCTTAGGTTAGTCGTACCGGTACAAGCTACTAGGCGGTTTGGGCTCTATTTGAGGCTGCGGGCCCTTTAGACGAATATCGTGGTCCGGCTTCCCAATTGCATTGTGTTGTACACGGACTGTCTACTGTCGAGAAGGAGAGCTCAAGAGCCGATAGGGGTTGTGCACACGGTTTTCGTTCGCTGGACTGGCTTCTAAGTCAACTGTGAGCCGGCTGGGGGCCGCTCACTCAGTCAAATCCCGATCCATTGATCGTCGTTTAGCGGCGGACGCTAGTTATCGCCAGACAGAGTCGGTCCTTAAGCCAGACTGATTCGTTCGGGCGGGGGATCAGCAACCAAGCGGCCCGGATGCGGACGTTTACCACCGAGCTATGTAATCCTAGCTCTCGGTTGGAGGCTTTTGAATTGCAAGGCGGGGTGGCTAAGCACGGCGAGCCTTCATCAATGCAATGTCAAGAGAGCAATGAGCACGACGGCGGTTGCCTCCCCCTGGTCTTTTACCGCTCAGGACTTAACTGAAATCTGATAGATGGCTAGTTGAGCTAGTCCAATCTTAAAACGAGCTCACCCTATGATTATTTCTGAGCCGCATTACCACCCATTCCTACAACATATCTTAACGTGTGGTGGCTCCGTCGATTTGCATTGCATCAGCCGTAGGGTTAAGTCTTACCGGGGTGGAGACAATGACGAGGCCGCCCTGCAAGGTTATTGACTGAGAAGCAGAGGGCGGTAAAGCTTATTAAGAAGGCGCAATTTTCTGTTTGGTATTTTGAGGGTTTTATAAAGCTTACAGTTGGCAAAGGGAGTATATGAATTAACCGGGACACAGCTCGCCGTCTTCGCCCTTTATAACGTGCCTGGTCGATGACTATTCGATACTGGGGCAAGAACGCGGTACCTAACATGCAGCGAGACGAACCGGTTAATACAAAGCACAACGGGCTCGTTATCTGAGGTGATTCGGCGATAAATAATATAGGAACAAACCAACTTTTTGACGGGACAGCCCTATCTATAATCTTTAAAGGCCAGGTCGGCCCAAGTCAGGCCAAGTAGATAATACTTCGTTTATGAGCGGGGGTGTCTGCATCGACCACCGCTTTGGTCTTCAGGGCTAATTGTATCTGTCACTTCTTACTCGAGAGTACCGAAAGATTTG"
```
:::
:::


# **Finding the Replication Origin, Part III**

Goals:

1\) Work through an initial window to find frequent k-mers meeting our threshold(t).

2\) Slide the window through the genome, update counts, and add frequent k-mers to candidate list.

3\) Slide window, update counts, and test using threshold again.


::: {.cell}

```{.r .cell-code}
initialize_k_mer_dict <- function(k){
  nucleotides <- c("A", "C", "G", "T")

  k_mers_dict <- expand.grid(rep(list(nucleotides), k)) %>%
    unite("k_mers", everything(), remove = TRUE, sep = "") %>%
    unique() %>%
    mutate(Count = 0)
  
  return(k_mers_dict)
}

clump_finding <- function(genome,L,k,t){
  #k = list of kmer
  initial_window<- str_sub(genome,1,L)
  kmer_dict <- initialize_k_mer_dict(k)
  ng <- nchar(genome)
  for(i in 1:(L-(k-1))){
    curr_kmer<- str_sub(initial_window,i,i+(k-1))
    curr_count<- count_pattern(initial_window,curr_kmer)
    curr_row <- which(kmer_dict$k_mers==curr_kmer)
    kmer_dict$Count[curr_row] <- curr_count
  }
  print("done")
  candidates <- kmer_dict %>%
    filter(Count >= t)%>%
    pull(k_mers)
  print("done")
  #goal:slide windows, update counts, and test using threshold again. 
  #currently have : my-dict contains Kmers+ counts through initial window. We also have our first set of candidates. 
  for(j in 2:(ng-(L-2))){
    rem_pattern <- str_sub(genome, j - 1, j - 1 + (k - 1))
    add_pattern <- str_sub(genome, j + (L - 2) - (k - 1), j + (L - 2))

    rem_row <- which(kmer_dict$k_mers == rem_pattern)
    add_row <- which(kmer_dict$k_mers == add_pattern)
    
    kmer_dict$Count[rem_row] <- kmer_dict$Count[rem_row] - 1
    kmer_dict$Count[add_row] <- kmer_dict$Count[add_row] + 1
    
    if(kmer_dict$Count[add_row] >= t){
      candidates <- append(candidates, add_pattern)
      candidates <- unique(candidates)
    }
  }
  return(candidates)
}
```
:::

::: {.cell}

```{.r .cell-code}
Ecoli<-scan("/Users/darvinojha/Desktop/E_coli.txt",what = "character", sep = NULL)
```
:::

::: {.cell}

```{.r .cell-code}
clump_finding(Ecoli,500,3,9)
```

::: {.cell-output .cell-output-stdout}
```
[1] "done"
[1] "done"
```
:::

::: {.cell-output .cell-output-stdout}
```
 [1] "AAA" "CAA" "TAA" "ACA" "CCA" "GCA" "TGA" "TTA" "AAC" "CAC" "ACC" "GGC"
[13] "TGC" "CAG" "AGG" "CTG" "GTG" "AAT" "CAT" "TCT" "GGT" "ATT" "TTT" "GCG"
[25] "TAC" "ATA" "GCC" "GAA" "CCG" "CGC" "CGA" "ATG" "TTG" "GGG" "CCC" "CGG"
[37] "TGG" "GAT" "GTT" "TCG" "ATC" "CGT" "GCT" "ACG" "GTC" "TCC" "TTC" "AAG"
[49] "TCA" "AGC" "GGA" "CTT" "TGT" "ACT" "CTC" "TAT" "AGT" "GAC" "GAG" "AGA"
[61] "TAG" "GTA" "CCT" "CTA"
```
:::
:::
